% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames,x11names}{xcolor}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\newsavebox\pandoc@box
\newcommand*\pandocbounded[1]{% scales image to fit in text height/width
  \sbox\pandoc@box{#1}%
  \Gscale@div\@tempa{\textheight}{\dimexpr\ht\pandoc@box+\dp\pandoc@box\relax}%
  \Gscale@div\@tempb{\linewidth}{\wd\pandoc@box}%
  \ifdim\@tempb\p@<\@tempa\p@\let\@tempa\@tempb\fi% select the smaller of both
  \ifdim\@tempa\p@<\p@\scalebox{\@tempa}{\usebox\pandoc@box}%
  \else\usebox{\pandoc@box}%
  \fi%
}
% Set default figure placement to htbp
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
% definitions for citeproc citations
\NewDocumentCommand\citeproctext{}{}
\NewDocumentCommand\citeproc{mm}{%
  \begingroup\def\citeproctext{#2}\cite{#1}\endgroup}
\makeatletter
 % allow citations to break across lines
 \let\@cite@ofmt\@firstofone
 % avoid brackets around text for \cite:
 \def\@biblabel#1{}
 \def\@cite#1#2{{#1\if@tempswa , #2\fi}}
\makeatother
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newenvironment{CSLReferences}[2] % #1 hanging-indent, #2 entry-spacing
 {\begin{list}{}{%
  \setlength{\itemindent}{0pt}
  \setlength{\leftmargin}{0pt}
  \setlength{\parsep}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1
   \setlength{\leftmargin}{\cslhangindent}
   \setlength{\itemindent}{-1\cslhangindent}
  \fi
  % set entry spacing
  \setlength{\itemsep}{#2\baselineskip}}}
 {\end{list}}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{\hfill\break\parbox[t]{\linewidth}{\strut\ignorespaces#1\strut}}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{\strut#1\strut}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{\strut#1\strut}}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}
% \usepackage{booktabs}
\usepackage{booktabs}
\usepackage{longtable}
\usepackage{array}
\usepackage{multirow}
\usepackage{wrapfig}
\usepackage{float}
\usepackage{colortbl}
\usepackage{pdflscape}
\usepackage{tabu}
\usepackage{threeparttable}
\usepackage{threeparttablex}
\usepackage[normalem]{ulem}
\usepackage{makecell}
\usepackage{xcolor}
\usepackage{bookmark}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={Causal Effect Estimation in Mendelian Randomisation Studies - Evaluating a Novel Bayesian Approach To Genetic Pleiotropy Versus Established Weighted Median Methodology},
  pdfauthor={B233241},
  colorlinks=true,
  linkcolor={Maroon},
  filecolor={Maroon},
  citecolor={Blue},
  urlcolor={blue},
  pdfcreator={LaTeX via pandoc}}

\title{Causal Effect Estimation in Mendelian Randomisation Studies - Evaluating a Novel Bayesian Approach To Genetic Pleiotropy Versus Established Weighted Median Methodology}
\author{B233241}
\date{September 2024 - June 2025}

\begin{document}
\maketitle

{
\hypersetup{linkcolor=}
\setcounter{tocdepth}{2}
\tableofcontents
}
\newpage

\subsection*{Acknowledgements}\label{acknowledgements}
\addcontentsline{toc}{subsection}{Acknowledgements}

I would like to acknowledge

\subsection*{Contributions}\label{contributions}
\addcontentsline{toc}{subsection}{Contributions}

Mine others

\subsection*{Statement of Originality}\label{statement-of-originality}
\addcontentsline{toc}{subsection}{Statement of Originality}

I confirm that all work is my own except where indicated, that all sources are clearly referenced\ldots.

\subsection*{Word Count}\label{word-count}
\addcontentsline{toc}{subsection}{Word Count}

Word count:
4053

\newpage

\section{Introduction and Background}\label{introduction-and-background}

\subsection{Introduction to Mendelian Randomisation (MR)}\label{introduction-to-mendelian-randomisation-mr}

Epidemiology is the study of determinants and distribution of disease across populations; a common epidemiological study aim is therefore to seek evidence as to whether a given exposure (e.g.~cigarette smoking) may cause a given outcome (e.g.~lung cancer)\textsuperscript{\citeproc{ref-coggon_chapter_2003}{1}}. Logistics limit experimental interventions across large groups, so insights into associations between exposures and outcomes are gleaned from observational data of people in the population of interest. Comparing health outcomes between individuals with different levels of a particular exposure may highlight potential links, e.g.~higher cancer incidence in those who smoke more is consistent with a causal role for cigarettes in carcinogenesis\textsuperscript{\citeproc{ref-coggon_chapter_2003}{1}}.

However, correlation does not prove causation. A key epidemiological challenge is accounting for so-called ``confounding'' factors; these are other variables, associated with both the exposure and the outcome of interest, which represent an alternative causal explanation for any exposure-outcome links observed\textsuperscript{\citeproc{ref-martens_instrumental_2006}{2}}. If smokers also drink more alcohol than non-smokers, then an observed link between smoking and increased cancer risk could plausibly be caused by increased alcohol exposure, either partially or entirely. Another potential issue with observational data is ``reverse causation'', where the presumed outcome is in fact a cause of the exposure; this might be the case if a cancer diagnosis drove individuals to drink and smoke more, and data were collected without respect to exposure timings.

\hyperref[acronyms_MR]{Mendelian randomisation (MR)} is a methodology intended to support causal inference from observational data. It applies the principles of \hyperref[acronyms_IV]{instrumental variable (IV)} analysis to genetic data, performing a type of natural experiment often likened to a \hyperref[acronyms_RCT]{randomised-controlled trial (RCT)}\textsuperscript{\citeproc{ref-hernan_instruments_2006}{3}}.

In a properly conducted \hyperref[acronyms_RCT]{RCT}, causality can be inferred due to a randomisation process being used as an ``instrument'' to allocate different levels of exposures to different experimental groups. If groups are randomly allocated, any confounding variables which might otherwise influence exposure-outcome relationships should be evenly distributed between groups, whether these confounders are known or not. As such, there should be no systematic differences between individuals from different groups in the exposure of interest - that is, there should be no bias\textsuperscript{\citeproc{ref-stel_instrumental_2013}{4}}. Statistical methods can quantify the probability that any observed outcome differences could have occurred by chance, and thereafter any outcome differences can be interpreted as caused by exposure differences. As allocation and receipt of exposures is known to precede outcome measurements, reverse causality is impossible.

In \hyperref[acronyms_MR]{MR}, naturally occurring genetic variants - ``genetic instruments'' -- are chosen based on their known association to an exposure of interest. Provided that assumptions of \hyperref[acronyms_IV]{IV} analysis are met, random assignment of alleles (i.e.~variants of a given gene) from parents to offspring during meiosis creates randomisation analagous to that performed for an \hyperref[acronyms_RCT]{RCT} -- both measured and unmeasured confounders should be distributed evenly between the groups created, allowing valid causal inference after other sources of bias and random variation are accounted for\textsuperscript{\citeproc{ref-davies_reading_2018}{5}}.

\subsection{Causal Effect Estimation in MR}\label{causal-effect-estimation-in-mr}

At its simplest, the relationship between two continuous variables - an exposure \(X\) and outcome \(Y\) - can be represented as a linear model:

\begin{equation} 
Y = \alpha + \beta X + \epsilon
\end{equation}

where \(\alpha\) represents all non-\(X\) determinants of \(Y\), \(\beta\) is the causal effect of \(X\) on \(Y\) and \(\epsilon\) is an error term. The \(\beta\) term is a numerical measure of strength of causal exposure-outcome association, where:

\begin{itemize}
\tightlist
\item
  \(\beta = 0\) implies no causal link between exposure and outcome
\item
  \(\beta > 0\) implies \(X\) causes \(Y\)
\item
  \(\beta < 0\) implies \(X\) prevents \(Y\)
\end{itemize}

To estimate a causal effect using a genetic variant in an \hyperref[acronyms_IV]{IV} analysis, three key assumptions must be met\textsuperscript{\citeproc{ref-lousdal_introduction_2018}{6}}:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Relevance -- the genetic variant must be associated with the exposure of interest
\item
  Independence -- the genetic variant is independent of confounders of the relationship between exposure and outcome
\item
  Exclusion restriction -- the genetic variant must not be associated with the outcome except via the exposure
\end{enumerate}

These assumptions are represented graphically in Figure \ref{fig:DAG-assumptions-plot}.

\begin{figure}
\centering
\pandocbounded{\includegraphics[keepaspectratio]{2_Intro_Background_files/figure-latex/DAG-assumptions-plot-1.pdf}}
\caption{\label{fig:DAG-assumptions-plot}Causal diagram illustrating the relationships between genetic instrument \emph{G}, exposure \emph{X}, outcome \emph{Y} and confounders of the exposure-outcome relationship \emph{U} in Mendelian randomisation studies. Blue text \& crosses represent key assumptions to ensure valid inference of causal effect of \emph{X} on \emph{Y} using \emph{G} as an instrumental variable. Red text represents violations of these assumptions that may lead to invalid inference through opening of alternate causal pathways. Greek characters represent the key parameters/association coefficients to be estimated. Adapted from Burgess et al 2016\textsuperscript{\citeproc{ref-burgess_sensitivity_2016}{7}}}
\end{figure}

Typically, \hyperref[acronyms_MR]{MR} studies estimate causal effect using a set of several genetic instruments; the causal effect estimate derived from the \(jth\) instrument is denoted \(\hat{\beta}_j\). Each estimate \(\hat{\beta}_j\) acknowledges there will be specific effects on the observed values of exposure and outcome given the presence of that specific genetic variable \(G_j\) under study, i.e.~\(\hat{\beta}_j\) is based on the instrument-conditioned exposure \({X|G_j}\) and outcome \({Y|G_j}\). These observed values of exposure and outcome can be described by their own linear models:

\begin{equation} 
X|G_j = \gamma_0 + \gamma_j G_j + \epsilon_{X_j}
\end{equation}

\begin{equation} 
Y|G_j = \Gamma_0 + \Gamma_j G_j + \epsilon_{Y_j}
\end{equation}

where, for exposure and outcome respectively:

\begin{itemize}
\tightlist
\item
  \(\gamma_0\) and \(\Gamma_0\) reflect base values without influence of the genetic variant
\item
  \(\gamma_j\) and \(\Gamma_j\) are coefficients of association with the genetic variant, representing the extent to which an effect allele of \(G_j\) will perturb the value of \(X\) or \(Y\) versus the non-effect allele
\item
  \(\epsilon_{X_j}\) and \(\epsilon_{Y_j}\) are error terms, containing contributions from confounders of the exposure-outcome relationship (\(U\) in the causal diagram), and all genetic variants except \(G_j\).
\end{itemize}

It can be shown that a simple causal effect estimate for the exposure on the outcome can be obtained from a single genetic instrument by the Wald method, dividing the coefficient of gene-outcome association by the coefficient of gene-exposure association, i.e.:

\begin{equation} 
\hat{\beta}_j = \frac {\hat{\Gamma}_j} {\hat{\gamma}_j}
\end{equation}

Each instrument may be valid or invalid, depending on it meeting the above assumptions. The overall causal effect estimate \(\hat{\beta}\) from any given \hyperref[acronyms_MR]{MR} method will typically seek to pool effect estimates from several instruments so as to minimise effects of any invalid instruments included, e.g.~by removing/down-weighting contributions of genetic instruments which violate one or more assumptions. This is equivalent to plotting all estimated coefficients of gene-outcome association (\(\bar{\Gamma}\)) versus all estimated coefficients of gene-exposure association (\(\bar{\gamma}\)) for the set of instruments, then using the gradient of a regression line through the points as the causal effect estimate \(\hat{\beta}\); picking an \hyperref[acronyms_MR]{MR} methodology is analogous to choosing the method to draw the line of best fit (Figure \ref{fig:Gamma-gamma-plot}). For binary outcomes, the causal effect estimate can be converted to an odds ratio (OR) through exponentiation, i.e.:

\begin{equation} 
OR = e^{\hat{\beta}}
\end{equation}

\begin{figure}
\centering
\pandocbounded{\includegraphics[keepaspectratio]{2_Intro_Background_files/figure-latex/Gamma-gamma-plot-1.pdf}}
\caption{\label{fig:Gamma-gamma-plot}Simulated MR Study on 10,000 individuals using 25 genetic instruments, of which 30\% are invalid (red points) and introduce directional pleiotropic effects. The true value of the exposure-outcome causal effect is 0.25 (grey line, causal effect represented by gradient). Regression using an unajusted least-squares linear model (light blue line) results in a biased estimate in the positive direction due to the influence of the invalid instruments. Using the Weighted Median Estimator method (pink line) attenuates the effects of the invalid instruments, resulting in an estimate closer to the true value. Adapted from Bowden et al 2016\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}}
\end{figure}

\subsection{Violations to Assumptions}\label{violations-to-assumptions}

In practice, only the relevance assumption can be directly tested and proven. Typically, genetic variants for \hyperref[acronyms_MR]{MR} studies are selected as instruments based on Genome Wide Association Studies (GWAS), which quantify associations between small genetic variations - known as \hyperref[acronyms_SNP]{single nucleotide polymorphism (SNP)}s - and various phenotypes. Association between genetic variants and a phenotypes representing exposures of interest can be partly assured by selection using an appropriate genome-wide significance level (e.g.~\(p < 10 ^{-8}\)). Statistical testing can also quantify the gene-exposure relationship; commonly used measures include the \(r^2\) statistic, representing the proportion of variance in the exposure explained by the genotype, and the related \(F\)-statistic, which additionally accounts for the the sample size under investigation\textsuperscript{\citeproc{ref-richmond_mendelian_2022}{9}}. An \(F\)-statistic of \(\ge\) 10 is generally considered to represent a strong enough gene-exposure association to consider a genetic instrument for use\textsuperscript{\citeproc{ref-martens_instrumental_2006}{2}}.

The assumptions of independence and exclusion restriction depend on all possible confounders of the exposure-outcome association, both measured and unmeasured; as such, these can never be proven absolutely. Various methods have been proposed to quantify and account for violations of these two additional assumptions, including the weighted median estimator, described below\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}.

The main methods to avoid violations of the independence assumption relate to appropriate selection of populations studied to avoid confounding due to ancestry or population stratification. For example, in two-sample MR studies, where gene-exposure and gene-outcome coefficients are estimated from two separate GWAS studies, it is recommended to select GWAS studies performed in similar population groups (e.g.~both in Western Europeans). This practice helps avoid spurious exposure-outcome associations being generated by confounding due to underlying differences in allele frequency, baseline disease risks etc between ancestrally different populations\textsuperscript{\citeproc{ref-richmond_mendelian_2022}{9}}.

Exclusion restriction is a particularly universal issue in MR, due to so-called (horizontal) genetic pleiotropy, where a single genetic variant may have multiple ``pleiotropic'' effects -- i.e.~it may influence several traits simultaneously. Such pleiotropic effects may be unknown and open unmeasured causal pathways between a genetic instrument and the outcome (Figure \ref{fig:DAG-assumptions-plot}), thus potentially biasing \hyperref[acronyms_MR]{MR} estimates of the association between exposure and outcome. As pleiotropy influences outcome separate to the path involving the exposure of interest, the term ``direct effects'' is also used\textsuperscript{\citeproc{ref-hemani_evaluating_2018}{10}}. Where pleiotropic effects are in both positive and negative directions with a mean of zero - ``balanced pleiotropy'' - then they only add noise to causal effect estimation\textsuperscript{\citeproc{ref-morrison_mendelian_2020}{11}}. By contrast, ``directional pleiotropy'', where the mean of pleiotropic effects is non-zero, may introduce bias\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}.

If such an additional causal pathway acts between gene \(G\) and outcome \(Y\) via a confounding factor \(U\), then the magnitude of direct/overall effects of \(G\) on \(Y\) will correlate with the effects of \(G\) on \(X\) (i.e.~\(\Gamma \propto \gamma\)), and ``correlated pleiotropy'' is present. If an additional causal pathway acts directly between gene \(G\) and outcome \(Y\) independent of both exposure \(X\) and confounders \(U\), this results in ``uncorrelated pleiotropy'' (Figure \ref{fig:DAG-assumptions-plot}). Both correlated and uncorrelated pleiotropy can introduce bias which distorts the estimate of the true causal effect. In general, correlated pleiotropy is more challenging to account for; several MR methods explicitly require an additional assumption of \hyperref[acronyms_InSIDE]{Instrument Strength Independent of Direct Effect (InSIDE)}, i.e no correlated pleiotropy to be present\textsuperscript{\citeproc{ref-grant_bayesian_2024}{12}}.

\subsection{Weighted Median Estimator (WME)}\label{weighted-median-estimator-wme}

A common approach to produce exposure-outcome causal effect estimates robust to violations of the exclusion restriction assumption is the \hyperref[acronyms_WME]{weighted median estimator (WME)} method, proposed by Bowden et al\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}.

In \hyperref[acronyms_WME]{WME} analysis, several genetic instruments are used to estimate the exposure-outcome causal effect \(\hat{\beta}\). Each instrument is known to be associated with the exposure of interest, but an unknown proportion of these instruments may be invalid due to pleiotropic genetic effects. Any instrument linked to an outcome via multiple pleiotropic causal pathways will exhibit a less consistent gene-outcome association than a relationship mediated by a single pathway; this results in larger variance in causal estimates derived from invalid/pleiotropic genetic instruments versus estimates from valid instruments.

\hyperref[acronyms_WME]{WME} therefore assigns a weight to each genetic instrument's estimate of the causal effect according to the inverse of the variance of the estimate; these weighted effect estimates are used to construct a cumulative distribution function for probability of true causal effect size across the range of estimated values. The 50th percentile of this distribution can then be taken as a ``weighted median estimate'' of the true causal effect, theoretically producing consistent causal estimates even if up to 50\% of the included information comes from invalid instruments\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}. An example of \hyperref[acronyms_WME]{WME} attenuating the effects of invalid instruments is shown in Figure \ref{fig:Gamma-gamma-plot}.

\subsection{Issues With WME CIs}\label{issues-with-wme-cis}

\hyperref[acronyms_WME]{WME} calculation methods are available via several prolific \hyperref[acronyms_MR]{MR} tools: the R packages ``MendelianRandomization''\textsuperscript{\citeproc{ref-yavorska_mendelianrandomization_2017}{13}} and ``TwoSampleMR'', and the MR-Base web platform\textsuperscript{\citeproc{ref-hemani_mr-base_2018}{14}}. However, these implement the original authors' suggested process of generating 95\% confidence intervals for \hyperref[acronyms_WME]{WME}, which deviates from accepted re-sampling methodology:

\begin{quote}
``We found the bootstrap confidence interval\ldots too conservative. However, the bootstrap standard error\ldots{} gave more reasonable coverage using a normal approximation (estimate ±1.96 x standard error) to form a 95\% confidence interval''\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}
\end{quote}

This modification, explicitly aiming to boost estimate precision artificially, would be expected to lead to a high Type 1 error rate, which has been a growing concern in the field of late\textsuperscript{\citeproc{ref-stender_reclaiming_2024}{15}}. The theoretical issues with this approach, and the fundamentals of bootstrapping in general, are covered in Appendix @(ref:appendix-boot).

\subsection{MR-Hevo}\label{mr-hevo}

MR-Hevo is an R package which uses more typical Bayesian methodology to estimate MR causal effects and corresponding 95\% confidence intervals. It uses the probabilistic programming language, Stan, to directly sample the posterior probability distribution of pleiotropic effects on the outcome, rather than making untested assumptions about the shape of this distribution as current \hyperref[acronyms_WME]{WME} implementations do\textsuperscript{\citeproc{ref-mckeigue_inference_2024}{16}}.

MR-Hevo incorporates several additional features which its creators claim further aid valid causal inference. Most MR methods can only account for one genetic variant per genetic locus (i.e.~per location in the genome). If multiple variants exist at a given locus, generally only one can be selected as an instrument for further \hyperref[acronyms_MR]{MR} analysis. HR-Hevo handles multiple instruments per genetic locus via scalar construction, essentially assigning a ``score'' to each locus based on the variant(s) present, thus incorporating more information than if closely grouped variants had been discarded\textsuperscript{\citeproc{ref-mckeigue_inference_2024}{16}}. As MR-Hevo is based on a Bayesian approach, it generates estimates which incorporate relevant existing information generated by prior studies, increasing the amount of data informing each estimate. In this case, MR-Hevo specifies a prior probability distribution reflecting prior knowledge that most individual genetic instruments will have only small effects on complex traits\textsuperscript{\citeproc{ref-park_estimation_2010}{17},\citeproc{ref-piironen_sparsity_2017}{18}}, further aiding biologically plausible inference regarding distribution of pleiotropic effects.

\subsection{Aims and Objectives}\label{aims-and-objectives}

The main aim of this study will be to demonstrate if the \hyperref[acronyms_WME]{WME} approach gives over-confident causal estimates in the presence of pleiotropy, and whether this issue is more correctly handled by the MR-Hevo Bayesian approach. This will be achieved through addressing the research questions and objectives as outlined below:

\subsubsection{Research Questions:}\label{research-questions}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  How does MR-Hevo perform versus the weighted median estimator when estimating causal effects in MR studies?
\item
  Do conclusions of existing MR studies using weighted median causal effect estimation change if MR-Hevo methods are used?
\end{enumerate}

\subsubsection{Objectives:}\label{objectives}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Quantify the precision of MR-Hevo causal estimates for simulated data under differing sets of common assumptions, with reference to the weighted median estimator
\item
  Evaluate the consistency of MR-Hevo causal estimates for simulated data under differing sets of common assumptions, with reference to the weighted median estimator
\item
  Compare the conclusions drawn from MR-Hevo causal effect estimation versus the weighted median estimator on real-world data
\end{enumerate}

\newpage

\section{Methods}\label{methods}

\subsection{Simulation Study}\label{simulation-study}

To evaluate the performance of MR-Hevo causal estimation relative to WME, the precision and consistency of both methods were quantified using simulated datasets with known parameter values.

\subsubsection{Data Simulation}\label{data-simulation}

To aid comparability with existing methods and literature, the simulation methodology of the original WME exposition was reproduced based on published models and parameters in Appendix 3 of its supplementary materials\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}. Full details of simulation reproduction, including code and validation of outputs, is presented in \ref{appendix-sim}.

In brief, simulations were created based on three different scenarios, each representing a common set of assumptions about underlying data used for MR, and each increasingly challenging to the performance of any given MR causal estimation methodology:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  Balanced pleiotropy, InSIDE assumption satisfied - A proportion of invalid genetic instruments are present and introduce pleiotropic effects uncorrelated with the instrument strength; these pleiotropic effects are equally likely to be positive as negative with a mean value = 0, thus introducing noise into the estimation of causal effect.
\item
  Directional pleiotropy, InSIDE assumption satisfied - A proportion of invalid genetic instruments are present and introduce pleiotropic effects uncorrelated with the instrument strength; these pleiotropic effects are positive only, with a mean value \textgreater{} 0, thus biasing the causal effect estimate in a positive direction.
\item
  Directional pleiotropy, InSIDE assumption not satisfied - A proportion of invalid genetic instruments are present and introduce pleiotropic effects correlated with the instrument strength through action via a confounder; these pleiotropic effects are positive only, with a mean value \textgreater{} 0, thus potentially biasing the causal effect estimate in a positive direction to an even greater extent than Scenario 2.
\end{enumerate}

1,000 simulated datasets of participant-level data were generated for every combination of each scenario and each the following simulation parameters:

\begin{itemize}
\tightlist
\item
  Proportion of invalid instruments: 0\%, 10\%, 20\% or 30\%
\item
  Number of participants: \(n = 10,000\) or \(n = 20,000\)
\item
  Causal effect: null (\(\beta = 0\)) or positive (\(\beta = 0.1\))
\end{itemize}

The same set of 25 simulated genetic instruments were used across all datasets, with the status of each as valid/invalid determined by random draw per instrument at the start of each simulation run of 1,000 datasets.

Genotypes were simulated as for a two-sample setting: where number of particpants was \(n = 10,000\), 20,000 genotypes were simulated - 10,000 for the cohort used to estimate gene-exposure association (\(\hat{\gamma}\)), and a separate cohort of 10,000 used to estimate gene-outcome association (\(\hat{\Gamma}\)). Parameter values for effect allele frequency were not specified by Bowden et al, though initial testing showed values around 0.5 produced WME causal effect estimates closest to published values when other parameters were matched\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}. As such, effect allele frequencies were assigned per instrument from a uniform distribution between 0.4 to 0.6. Each effect allele freqency thus generated per instrument was then used as a probability to assign each simulated participant effect alleles for each instrument via two draws from a binomial distribution.

\subsubsection{Analysis of Simulated Data}\label{analysis-of-simulated-data}

Each dataset generated was analysed using both WME and MR-Hevo methods, via functions from the \texttt{TwoSampleMR} and \texttt{mrhevo} packages, respectively\textsuperscript{\citeproc{ref-hemani_mr-base_2018}{14},\citeproc{ref-mckeigue_inference_2024}{16}}. Results were aggregated per group of 1,000 simulated datasets corresponding to a particular combination of scenario and parameter values. This resulted in one meta-analysis reported per combination of scenario/parameter values, each including 1,000 simulated MR studies using the same 25 genetic instruments in the same population. Aggregated measures for both WME and MR-Hevo per meta-analysis were mean causal effect estimate; mean standard error of the causal effect estimate; and causality report rate, i.e.~percentage of simulated studies reported as showing a non-null causal effect, either by p-value \textless0.05 (WME), or by a 95\% credible interval for causal effect estimate not including 0 (MR-Hevo).

Results of the above aggregations were tabulated as per Tables 2 and 3 of Bowden et al\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}} to allow direct comparisons of both methods versus each other and versus the published characteristics of existing MR causal estimation methods.

\subsection{Re-Analysis of Published Data}\label{re-analysis-of-published-data}

To investigate the potential implications of any differences in performance between WME and MR-Hevo methods, a selection of published studies resporting causal effect estimates using the WME method was re-analysed. A sample size of 10 published studies was decided as a pragmatic compromise between the scope of this study and the need to check consistency of any observed differences. In the original Bowden et al simulation studies, the WME causal estimation method was shown to generate a false-positive report rate of \(\ge\) 30\% with relatively minor violations of relevant assumptions\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}; therefore, even this relatively small sample of 10 studies might be expected to demonstrate differences between methods if the MR-Hevo approach is as appropriately conservative as its creators propose.

To estimate the upper bound of the potential impact of MR-Hevo versus existing WME methodology, studies were chosen for re-analysis based on their number of citations in the wider MR literature. Compared to studies with few or no citations, highly-cited studies would be expected to have a larger impact on their respective fields if their conclusions were to change. In addition, highly-cited works will typically have been submitted to more scrutiny than less-cited works - both during peer review whilst under consideration by journals likely to produce highly-cited works, and from the wider scientific community following the widespread dissemination evidenced by a high citation count. As such, it would be expected that highly-cited works are likely to be free of significant methodological flaws which may impede interpretation of any re-analysis.

\subsubsection{Citation Search}\label{citation-search}

The Scopus search platform {[}@{]} was used on 15/04/2025 to retrieve all articles citing the original weighted median estimator exposition paper\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}. The articles returned were sorted by the number of times each article itself had been cited, and the resulting list was saved to RIS format in blocks of ten articles for upload into the Covidence evidence synthesis platform. Abstracts were screened by a single reviewer (B233241), starting with the most cited article and proceeding in descending order of citation count, against the following inclusion and exclusion criteria:

Inclusion criteria:

\begin{itemize}
\item
  Original two-sample MR study
\item
  Able to determine samples' ancestry sufficient to establish presence/potential degree of participant overlap between groups
\item
  Reporting \(\ge\) 20 human genetic instruments relating to exposure
\item
  Reports details of effect/non-effect alleles
\item
  Regression coefficients and standard errors and/or confidence intervals available for each genetic instrument used
\item
  Uses Weighted Median Estimator
\end{itemize}

Exclusion criteria:

\begin{itemize}
\item
  Methodology paper, review article, editorial or letter
\item
  English full-text not accessible
\end{itemize}

Where eligibility could not be determined from abstract screening alone, full texts were retrieved and screened against the same criteria. Screening of abstracts and full texts was undertaken in blocks of ten articles, until the target of ten included studies for reanalysis had been reached.

Where an article reported multiple exposure-outcome associations, data were only extracted for the association with the highest number of genetic instruments available, or else for the first reported association where several were based on the same number of instruments. Data were extracted from full texts of included studies using a standardised data collection template, which included publication details, citation count, primary study question, degree of participant overlap between groups, number/details of genetic instruments used, effect estimates/standard errors calculated, and conclusion regarding causality as determined by the weighted median estimator method.

\subsection{Data Manipulation and Analysis}\label{data-manipulation-and-analysis}

All simulations, data manipulations and data analyses were performed in R version 4.4.3 (2025-02-28 ucrt)\textsuperscript{\citeproc{ref-base}{19}}.

For the simulation study, full details of computation are available in Appendix \ref{appendix-sim}.

For citation search data, a standardised data collection form was Microsoft Excel\textsuperscript{\citeproc{ref-microsoft_corporation_microsoft_2018}{20}} to create .csv files for subsequent analysis in R; Excel's ``Get Data'' function was also used to extract tables of genetic instruments where these were presented in non-csv format (e.g.~pdf).

Data cleaning for citation search data was primarily undertaken using the Tidyverse suite of R packages\textsuperscript{\citeproc{ref-tidyverse}{21}}. A full list of packages used can be found in Appendix @(ref:appendix-pkg).

Data were manually screened at summary level and relevant features were extracted. Data were checked for completeness, consistency, duplicate values and plausibility. Data were transformed to an appropriate data type, and encoding of genetic variables was standardised into a single format. Missing values for association coefficients and \hyperref[acronyms_SE]{standard error (SE)}s were imputed as the mean value calculated per dataset. It was noted during early testing that MR-Hevo functions do not operate correctly when zero values are present in coefficients of genetic association or their standard errors; such zero values were therefore re-coded as an arbitrarily low value of \(10^{-100}\).

\subsection{Ethical Approval}\label{ethical-approval}

The protocol for this work has been reviewed and approved by the \hyperref[acronyms_UMREG]{Usher Masters Research Ethics Group (UMREG)} at the University of Edinburgh, Ethics ID: UM241126. Due to the nature of the project, using simulated and publically available data only, no significant ethical issues were foreseen, and sponsorship was deemed unnecessary by the \hyperref[acronyms_UMREG]{UMREG} reviewing panel.

\section{References}\label{references}

\phantomsection\label{refs}
\begin{CSLReferences}{0}{1}
\bibitem[\citeproctext]{ref-coggon_chapter_2003}
\CSLLeftMargin{1. }%
\CSLRightInline{Coggon D, Rose G, Barker D. Chapter 1. {What} is epidemiology? {\textbar} {The} {BMJ}. In: The {BMJ} {\textbar} {The} {BMJ}: Leading general medical journal {Research} {Education} {Comment} {[}Internet{]}. 2003 {[}cited 2025 Apr 29{]}. Available from: \url{https://www.bmj.com/about-bmj/resources-readers/publications/epidemiology-uninitiated/1-what-epidemiology}}

\bibitem[\citeproctext]{ref-martens_instrumental_2006}
\CSLLeftMargin{2. }%
\CSLRightInline{Martens EP, Pestman WR, Boer A de, Belitser SV, Klungel OH. Instrumental {Variables}: {Application} and {Limitations}. Epidemiology {[}Internet{]}. 2006 May {[}cited 2025 Apr 29{]};17(3):260. Available from: \url{https://journals.lww.com/epidem/fulltext/2006/05000/instrumental_variables__application_and.10.aspx}}

\bibitem[\citeproctext]{ref-hernan_instruments_2006}
\CSLLeftMargin{3. }%
\CSLRightInline{Hernán MA, Robins JM. Instruments for {Causal} {Inference}: {An} {Epidemiologist}'s {Dream}? Epidemiology {[}Internet{]}. 2006 Jul {[}cited 2025 Apr 29{]};17(4):360. Available from: \url{https://journals.lww.com/epidem/fulltext/2006/07000/instruments_for_causal_inference__an.4.aspx\#JCL1-2}}

\bibitem[\citeproctext]{ref-stel_instrumental_2013}
\CSLLeftMargin{4. }%
\CSLRightInline{Stel VS, Dekker FW, Zoccali C, Jager KJ. Instrumental variable analysis. Nephrology Dialysis Transplantation {[}Internet{]}. 2013 Jul {[}cited 2025 Apr 29{]};28(7):1694--9. Available from: \url{https://doi.org/10.1093/ndt/gfs310}}

\bibitem[\citeproctext]{ref-davies_reading_2018}
\CSLLeftMargin{5. }%
\CSLRightInline{Davies NM, Holmes MV, Smith GD. Reading {Mendelian} randomisation studies: A guide, glossary, and checklist for clinicians. BMJ {[}Internet{]}. 2018 Jul {[}cited 2025 Jan 7{]};362:k601. Available from: \url{https://www.bmj.com/content/362/bmj.k601}}

\bibitem[\citeproctext]{ref-lousdal_introduction_2018}
\CSLLeftMargin{6. }%
\CSLRightInline{Lousdal ML. An introduction to instrumental variable assumptions, validation and estimation. Emerging Themes in Epidemiology {[}Internet{]}. 2018 Jan {[}cited 2025 Apr 29{]};15:1. Available from: \url{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5776781/}}

\bibitem[\citeproctext]{ref-burgess_sensitivity_2016}
\CSLLeftMargin{7. }%
\CSLRightInline{Burgess S, Bowden J, Fall T, Ingelsson E, Thompson SG. Sensitivity {Analyses} for {Robust} {Causal} {Inference} from {Mendelian} {Randomization} {Analyses} with {Multiple} {Genetic} {Variants}. Epidemiology (Cambridge, Mass) {[}Internet{]}. 2016 Nov {[}cited 2024 Oct 22{]};28(1):30. Available from: \url{https://pmc.ncbi.nlm.nih.gov/articles/PMC5133381/}}

\bibitem[\citeproctext]{ref-bowden_consistent_2016}
\CSLLeftMargin{8. }%
\CSLRightInline{Bowden J, Smith GD, Haycock PC, Burgess S. Consistent {Estimation} in {Mendelian} {Randomization} with {Some} {Invalid} {Instruments} {Using} a {Weighted} {Median} {Estimator}. Genetic Epidemiology {[}Internet{]}. 2016 Apr {[}cited 2024 Oct 22{]};40(4):304. Available from: \url{https://pmc.ncbi.nlm.nih.gov/articles/PMC4849733/}}

\bibitem[\citeproctext]{ref-richmond_mendelian_2022}
\CSLLeftMargin{9. }%
\CSLRightInline{Richmond RC, Smith GD. Mendelian {Randomization}: {Concepts} and {Scope}. Cold Spring Harbor Perspectives in Medicine {[}Internet{]}. 2022 Jan {[}cited 2024 Oct 22{]};12(1):a040501. Available from: \url{https://pmc.ncbi.nlm.nih.gov/articles/PMC8725623/}}

\bibitem[\citeproctext]{ref-hemani_evaluating_2018}
\CSLLeftMargin{10. }%
\CSLRightInline{Hemani G, Bowden J, Smith GD. Evaluating the potential role of pleiotropy in {Mendelian} randomization studies. Human Molecular Genetics {[}Internet{]}. 2018 May {[}cited 2024 Oct 23{]};27(R2):R195. Available from: \url{https://pmc.ncbi.nlm.nih.gov/articles/PMC6061876/}}

\bibitem[\citeproctext]{ref-morrison_mendelian_2020}
\CSLLeftMargin{11. }%
\CSLRightInline{Morrison J, Knoblauch N, Marcus JH, Stephens M, He X. Mendelian randomization accounting for correlated and uncorrelated pleiotropic effects using genome-wide summary statistics. Nature Genetics {[}Internet{]}. 2020 Jul {[}cited 2025 May 22{]};52(7):740--7. Available from: \url{https://www.nature.com/articles/s41588-020-0631-4}}

\bibitem[\citeproctext]{ref-grant_bayesian_2024}
\CSLLeftMargin{12. }%
\CSLRightInline{Grant AJ, Burgess S. A {Bayesian} approach to {Mendelian} randomization using summary statistics in the univariable and multivariable settings with correlated pleiotropy. The American Journal of Human Genetics {[}Internet{]}. 2024 Jan {[}cited 2025 May 20{]};111(1):165--80. Available from: \url{https://www.cell.com/ajhg/abstract/S0002-9297(23)00433-0}}

\bibitem[\citeproctext]{ref-yavorska_mendelianrandomization_2017}
\CSLLeftMargin{13. }%
\CSLRightInline{Yavorska OO, Burgess S. {MendelianRandomization}: An {R} package for performing {Mendelian} randomization analyses using summarized data. International Journal of Epidemiology {[}Internet{]}. 2017 Dec {[}cited 2024 Oct 23{]};46(6):1734--9. Available from: \url{https://doi.org/10.1093/ije/dyx034}}

\bibitem[\citeproctext]{ref-hemani_mr-base_2018}
\CSLLeftMargin{14. }%
\CSLRightInline{Hemani G, Zheng J, Elsworth B, Wade KH, Haberland V, Baird D, et al. The {MR}-{Base} platform supports systematic causal inference across the human phenome. Loos R, editor. eLife {[}Internet{]}. 2018 May {[}cited 2025 Jan 7{]};7:e34408. Available from: \url{https://doi.org/10.7554/eLife.34408}}

\bibitem[\citeproctext]{ref-stender_reclaiming_2024}
\CSLLeftMargin{15. }%
\CSLRightInline{Stender S, Gellert-Kristensen H, Smith GD. Reclaiming mendelian randomization from the deluge of papers and misleading findings. Lipids in Health and Disease {[}Internet{]}. 2024 Sep {[}cited 2025 Jan 7{]};23(1):286. Available from: \url{https://doi.org/10.1186/s12944-024-02284-w}}

\bibitem[\citeproctext]{ref-mckeigue_inference_2024}
\CSLLeftMargin{16. }%
\CSLRightInline{McKeigue PM, Iakovliev A, Spiliopoulou A, Erabadda B, Colhoun HM. Inference of causal and pleiotropic effects with multiple weak genetic instruments: Application to effect of adiponectin on type 2 diabetes {[}Internet{]}. medRxiv; 2024 {[}cited 2024 Oct 23{]}. Available from: \url{https://www.medrxiv.org/content/10.1101/2023.12.15.23300008v2}}

\bibitem[\citeproctext]{ref-park_estimation_2010}
\CSLLeftMargin{17. }%
\CSLRightInline{Park JH, Wacholder S, Gail MH, Peters U, Jacobs KB, Chanock SJ, et al. Estimation of effect size distribution from genome-wide association studies and implications for future discoveries. Nature genetics {[}Internet{]}. 2010 Jun {[}cited 2024 Oct 23{]};42(7):570. Available from: \url{https://pmc.ncbi.nlm.nih.gov/articles/PMC4615599/}}

\bibitem[\citeproctext]{ref-piironen_sparsity_2017}
\CSLLeftMargin{18. }%
\CSLRightInline{Piironen J, Vehtari A. Sparsity information and regularization in the horseshoe and other shrinkage priors. Electronic Journal of Statistics {[}Internet{]}. 2017 Jan {[}cited 2025 Jan 7{]};11(2):5018--51. Available from: \url{https://projecteuclid.org/journals/electronic-journal-of-statistics/volume-11/issue-2/Sparsity-information-and-regularization-in-the-horseshoe-and-other-shrinkage/10.1214/17-EJS1337SI.full}}

\bibitem[\citeproctext]{ref-base}
\CSLLeftMargin{19. }%
\CSLRightInline{R Core Team. {R}: A language and environment for statistical computing {[}Internet{]}. Vienna, Austria: R Foundation for Statistical Computing; 2025. Available from: \url{https://www.R-project.org/}}

\bibitem[\citeproctext]{ref-microsoft_corporation_microsoft_2018}
\CSLLeftMargin{20. }%
\CSLRightInline{Microsoft Corporation. Microsoft {Excel} {[}Internet{]}. 2018. Available from: \url{https://office.microsoft.com/excel}}

\bibitem[\citeproctext]{ref-tidyverse}
\CSLLeftMargin{21. }%
\CSLRightInline{Wickham H, Averick M, Bryan J, Chang W, McGowan LD, François R, et al. \href{https://doi.org/10.21105/joss.01686}{Welcome to the {tidyverse}}. Journal of Open Source Software. 2019;4(43):1686. }

\bibitem[\citeproctext]{ref-buscaglia_chapter_2020}
\CSLLeftMargin{22. }%
\CSLRightInline{Buscaglia DLS\&R. Chapter 3 {Confidence} {Intervals} via {Bootstrapping} {\textbar} {Introduction} to {Statistical} {Methodology}, {Second} {Edition} {[}Internet{]}. 2020 {[}cited 2025 May 25{]}. Available from: \url{https://bookdown.org/dereksonderegger/570/3-confidence-intervals-via-bootstrapping.html}}

\bibitem[\citeproctext]{ref-ross_chapter_2014}
\CSLLeftMargin{23. }%
\CSLRightInline{Ross SM. Chapter 6 - {Distributions} of {Sampling} {Statistics}. In: Ross SM, editor. Introduction to {Probability} and {Statistics} for {Engineers} and {Scientists} ({Fifth} {Edition}) {[}Internet{]}. Boston: Academic Press; 2014 {[}cited 2025 May 25{]}. p. 207--33. Available from: \url{https://www.sciencedirect.com/science/article/pii/B978012394811350006X}}

\bibitem[\citeproctext]{ref-cata_blood_2011}
\CSLLeftMargin{24. }%
\CSLRightInline{Cata JP, Klein EA, Hoeltge GA, Dalton JE, Mascha E, O'Hara J, et al. Blood {Storage} {Duration} and {Biochemical} {Recurrence} of {Cancer} {After} {Radical} {Prostatectomy}. Mayo Clinic Proceedings {[}Internet{]}. 2011 Feb {[}cited 2025 May 25{]};86(2):120--7. Available from: \url{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3031436/}}

\bibitem[\citeproctext]{ref-medicaldata}
\CSLLeftMargin{25. }%
\CSLRightInline{Higgins P. {medicaldata}: Data package for medical datasets {[}Internet{]}. 2021. Available from: \url{https://CRAN.R-project.org/package=medicaldata}}

\bibitem[\citeproctext]{ref-acronymsdown}
\CSLLeftMargin{26. }%
\CSLRightInline{Chaput R. {acronymsdown}: Acronyms and glossaries support for RMarkdown {[}Internet{]}. 2025. Available from: \url{https://github.com/rchaput/acronymsdown}}

\bibitem[\citeproctext]{ref-bookdown2016}
\CSLLeftMargin{27. }%
\CSLRightInline{Xie Y. {bookdown}: Authoring books and technical documents with {R} markdown {[}Internet{]}. Boca Raton, Florida: Chapman; Hall/CRC; 2016. Available from: \url{https://bookdown.org/yihui/bookdown}}

\bibitem[\citeproctext]{ref-bookdown2025}
\CSLLeftMargin{28. }%
\CSLRightInline{Xie Y. {bookdown}: Authoring books and technical documents with r markdown {[}Internet{]}. 2025. Available from: \url{https://github.com/rstudio/bookdown}}

\bibitem[\citeproctext]{ref-car}
\CSLLeftMargin{29. }%
\CSLRightInline{Fox J, Weisberg S. An {R} companion to applied regression {[}Internet{]}. Third. Thousand Oaks {CA}: Sage; 2019. Available from: \url{https://www.john-fox.ca/Companion/}}

\bibitem[\citeproctext]{ref-cowplot}
\CSLLeftMargin{30. }%
\CSLRightInline{Wilke CO. {cowplot}: Streamlined plot theme and plot annotations for {``{ggplot2}''} {[}Internet{]}. 2024. Available from: \url{https://CRAN.R-project.org/package=cowplot}}

\bibitem[\citeproctext]{ref-crayon}
\CSLLeftMargin{31. }%
\CSLRightInline{Csárdi G. {crayon}: Colored terminal output {[}Internet{]}. 2024. Available from: \url{https://CRAN.R-project.org/package=crayon}}

\bibitem[\citeproctext]{ref-devtools}
\CSLLeftMargin{32. }%
\CSLRightInline{Wickham H, Hester J, Chang W, Bryan J. {devtools}: Tools to make developing r packages easier {[}Internet{]}. 2022. Available from: \url{https://CRAN.R-project.org/package=devtools}}

\bibitem[\citeproctext]{ref-ggdag}
\CSLLeftMargin{33. }%
\CSLRightInline{Barrett M. {ggdag}: Analyze and create elegant directed acyclic graphs {[}Internet{]}. 2024. Available from: \url{https://CRAN.R-project.org/package=ggdag}}

\bibitem[\citeproctext]{ref-gghighlight}
\CSLLeftMargin{34. }%
\CSLRightInline{Yutani H. {gghighlight}: Highlight lines and points in {``{ggplot2}''} {[}Internet{]}. 2023. Available from: \url{https://CRAN.R-project.org/package=gghighlight}}

\bibitem[\citeproctext]{ref-grateful}
\CSLLeftMargin{35. }%
\CSLRightInline{Rodriguez-Sanchez F, Jackson CP. {grateful}: Facilitate citation of {R} packages {[}Internet{]}. 2024. Available from: \url{https://pakillo.github.io/grateful/}}

\bibitem[\citeproctext]{ref-grid}
\CSLLeftMargin{36. }%
\CSLRightInline{R Core Team. {R}: A language and environment for statistical computing {[}Internet{]}. Vienna, Austria: R Foundation for Statistical Computing; 2025. Available from: \url{https://www.R-project.org/}}

\bibitem[\citeproctext]{ref-here}
\CSLLeftMargin{37. }%
\CSLRightInline{Müller K. {here}: A simpler way to find your files {[}Internet{]}. 2020. Available from: \url{https://CRAN.R-project.org/package=here}}

\bibitem[\citeproctext]{ref-infer}
\CSLLeftMargin{38. }%
\CSLRightInline{Couch SP, Bray AP, Ismay C, Chasnovski E, Baumer BS, Çetinkaya-Rundel M. \href{https://doi.org/10.21105/joss.03661}{{infer}: An {R} package for tidyverse-friendly statistical inference}. Journal of Open Source Software. 2021;6(65):3661. }

\bibitem[\citeproctext]{ref-kableExtra}
\CSLLeftMargin{39. }%
\CSLRightInline{Zhu H. {kableExtra}: Construct complex table with {``{kable}''} and pipe syntax {[}Internet{]}. 2024. Available from: \url{https://CRAN.R-project.org/package=kableExtra}}

\bibitem[\citeproctext]{ref-knitr2014}
\CSLLeftMargin{40. }%
\CSLRightInline{Xie Y. {knitr}: A comprehensive tool for reproducible research in {R}. In: Stodden V, Leisch F, Peng RD, editors. Implementing reproducible computational research. Chapman; Hall/CRC; 2014. }

\bibitem[\citeproctext]{ref-knitr2015}
\CSLLeftMargin{41. }%
\CSLRightInline{Xie Y. Dynamic documents with {R} and knitr {[}Internet{]}. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC; 2015. Available from: \url{https://yihui.org/knitr/}}

\bibitem[\citeproctext]{ref-knitr2025}
\CSLLeftMargin{42. }%
\CSLRightInline{Xie Y. {knitr}: A general-purpose package for dynamic report generation in {R} {[}Internet{]}. 2025. Available from: \url{https://yihui.org/knitr/}}

\bibitem[\citeproctext]{ref-matrixStats}
\CSLLeftMargin{43. }%
\CSLRightInline{Bengtsson H. {matrixStats}: Functions that apply to rows and columns of matrices (and to vectors) {[}Internet{]}. 2025. Available from: \url{https://CRAN.R-project.org/package=matrixStats}}

\bibitem[\citeproctext]{ref-parallel}
\CSLLeftMargin{44. }%
\CSLRightInline{R Core Team. {R}: A language and environment for statistical computing {[}Internet{]}. Vienna, Austria: R Foundation for Statistical Computing; 2025. Available from: \url{https://www.R-project.org/}}

\bibitem[\citeproctext]{ref-rmarkdown2018}
\CSLLeftMargin{45. }%
\CSLRightInline{Xie Y, Allaire JJ, Grolemund G. R markdown: The definitive guide {[}Internet{]}. Boca Raton, Florida: Chapman; Hall/CRC; 2018. Available from: \url{https://bookdown.org/yihui/rmarkdown}}

\bibitem[\citeproctext]{ref-rmarkdown2020}
\CSLLeftMargin{46. }%
\CSLRightInline{Xie Y, Dervieux C, Riederer E. R markdown cookbook {[}Internet{]}. Boca Raton, Florida: Chapman; Hall/CRC; 2020. Available from: \url{https://bookdown.org/yihui/rmarkdown-cookbook}}

\bibitem[\citeproctext]{ref-rmarkdown2024}
\CSLLeftMargin{47. }%
\CSLRightInline{Allaire J, Xie Y, Dervieux C, McPherson J, Luraschi J, Ushey K, et al. {rmarkdown}: Dynamic documents for r {[}Internet{]}. 2024. Available from: \url{https://github.com/rstudio/rmarkdown}}

\bibitem[\citeproctext]{ref-rstan}
\CSLLeftMargin{48. }%
\CSLRightInline{Stan Development Team. {RStan}: The {R} interface to {Stan} {[}Internet{]}. 2025. Available from: \url{https://mc-stan.org/}}

\bibitem[\citeproctext]{ref-TwoSampleMR2017}
\CSLLeftMargin{49. }%
\CSLRightInline{Hemani G, Tilling K, Davey Smith G. Orienting the causal relationship between imprecisely measured traits using GWAS summary data. PLoS Genetics {[}Internet{]}. 2017;13(11):e1007081. Available from: \url{https://doi.org/10.1371/journal.pgen.1007081}}

\bibitem[\citeproctext]{ref-TwoSampleMR2018}
\CSLLeftMargin{50. }%
\CSLRightInline{Hemani G, Zheng J, Elsworth B, Wade K, Baird D, Haberland V, et al. The MR-base platform supports systematic causal inference across the human phenome. eLife {[}Internet{]}. 2018;7:e34408. Available from: \url{https://elifesciences.org/articles/34408}}

\bibitem[\citeproctext]{ref-wordcountaddin}
\CSLLeftMargin{51. }%
\CSLRightInline{Marwick B. {wordcountaddin}: Word counts and readability statistics in r markdown documents {[}Internet{]}. 2025. Available from: \url{https://github.com/benmarwick/wordcountaddin}}

\end{CSLReferences}

\newpage

\appendix


\newpage

\section{Appendix: List of Abbreviations}\label{appendix-acr}

\begin{description}
\tightlist
\item[\phantomsection\label{acronyms_CI}{CI}]
confidence interval
\item[\phantomsection\label{acronyms_CLT}{CLT}]
central limit theorem
\item[\phantomsection\label{acronyms_IV}{IV}]
instrumental variable
\item[\phantomsection\label{acronyms_InSIDE}{InSIDE}]
Instrument Strength Independent of Direct Effect
\item[\phantomsection\label{acronyms_MR}{MR}]
Mendelian randomisation
\item[\phantomsection\label{acronyms_RCT}{RCT}]
randomised-controlled trial
\item[\phantomsection\label{acronyms_SD}{SD}]
standard deviation
\item[\phantomsection\label{acronyms_SE}{SE}]
standard error
\item[\phantomsection\label{acronyms_SNP}{SNP}]
single nucleotide polymorphism
\item[\phantomsection\label{acronyms_UMREG}{UMREG}]
Usher Masters Research Ethics Group
\item[\phantomsection\label{acronyms_WME}{WME}]
weighted median estimator
\end{description}

\newpage

\section{Appendix: Bootstrapping}\label{appendix-boot}

\subsection{Bootstrapping - General Method}\label{bootstrapping---general-method}

The typical process for ``bootstrap'' generating an estimate, \hyperref[acronyms_SE]{SE} and \hyperref[acronyms_CI]{confidence interval (CI)}s of a population parameter (e.g.~population mean \(\mu\)) from a sample \(x\) is as follows\textsuperscript{\citeproc{ref-buscaglia_chapter_2020}{22}}:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  A sample, \(x\), of \(n\) individuals is selected from a total population, \(X\), of \(N\) individuals
\item
  This sample \(x\) is then treated as the ``bootstrap population''; the empirical distribution of values in the \(n\) individuals in the bootstrap population is taken to be broadly representative of the distribution of values in the underlying population \(X\) of \(N\) individuals
\item
  A ``bootstrap sample'', \(x^*\), is then obtained by re-sampling individuals from the bootstrap population with replacement \(n\) times per bootstrap sample, i.e.~the new bootstrap sample comprises \(n\) sampled individuals, \(x^*_1, x^*_2,...x^*_n\). As such, individuals from the original bootstrap population \(x\) may contribute once, more than once or not at all to each bootstrap sample \(x^*\).
\item
  A total of \(k\) bootstrap samples are generated, \(x^{*1}, x^{*2},...x^{*k}\), and the statistic of interest (e.g.~sample mean \(\bar{x}\)) is estimated in each individual sample, \(\bar{x}^{*i}\), giving the complete set of \(\bar{x}^{*1}, \bar{x}^{*2},...\bar{x}^{*i}...\bar{x}^{*k}\).
\item
  The set of \(k\) statistics are combined to form a ``bootstrap distribution''; as expected from \hyperref[acronyms_CLT]{central limit theorem (CLT)}\textsuperscript{\citeproc{ref-ross_chapter_2014}{23}}, this is typically closer to a normal distribution than the underlying distribution of values in either the bootstrap population \(x\) or the total population \(X\). (See Figure \ref{fig:prostate-vol} for an example of this)
\item
  The final values are derived as follows:
\end{enumerate}

\begin{quote}
\begin{itemize}
\tightlist
\item
  the parameter estimate (e.g.~estimate of the true population mean, \(\hat{\mu}\)) is taken as the mean of the bootstrap distribution of \(k\) estimates, \((\sum^k_{i = 1} \bar{x}) \div n\)
\item
  the \hyperref[acronyms_CI]{CI}s are taken as the values at the appropriate centiles at the edges of the sampling distribution, e.g.~a 95\% \hyperref[acronyms_CI]{CI} would be generated using values at the 2.5th and 97.5th centiles
\item
  the \hyperref[acronyms_SE]{SE} of the estimate is taken as the \hyperref[acronyms_SD]{standard deviation (SD)} of the sampling distribution, given by \(\sqrt{\frac{1}{k - 1} \sum^k_{i = 1} (\bar{x_i} - \hat{\mu})^2}\)
\end{itemize}
\end{quote}

\subsection{Bootstrapping - Example: Prostate Volume}\label{bootstrapping---example-prostate-volume}

The above process is illustrated in \ref{fig:prostate-vol}. Data on prostate volume in 307 prostate cancer patients demonstrates a right-skewed distribution (A). An empirical distribution from a sample of 100 of these patients mirrors this right skew, and is used as the ``bootstrap population'' (B) for further re-sampling. As the bootstrap population is re-sampled more and more times, the ``bootstrap distribution'' of the sample means generated (C and D) gradually tends towards a normal distribution. The 95\% \hyperref[acronyms_CI]{CI} is given by the bounds defining the middle 95\% of the bootstrap distribution of estimated means, as shown.

\begin{figure}
\centering
\pandocbounded{\includegraphics[keepaspectratio]{9_Appendices_files/figure-latex/prostate-vol-1.pdf}}
\caption{\label{fig:prostate-vol}Histograms demonstrating distribution of prostate volumes in patients with prostatic cancer, taken from Cata et al 2011\textsuperscript{\citeproc{ref-cata_blood_2011}{24}} via the R package \texttt{medicaldata}\textsuperscript{\citeproc{ref-medicaldata}{25}}. A) Distribution from whole study population of 307 patients with non-missing data, exhibiting right-skew. B) Distribution from random sample of 100 patients, still exhibiting right-skew. C) Bootstrap distribution generated by re-sampling 1,000 bootstrap samples from the original sample of 100 patients, right-skew less apparent. D) Bootstrap distribution generated by re-sampling 100,000 bootstrap samples from the original sample of 100 patients, approaching normality. 95\% confidence intervals are demonstrated in plots C and D by marking the 2.5th and 97.5th centiles.}
\end{figure}

\newpage

\subsection{Bootstrapping - Relevance to WME}\label{bootstrapping---relevance-to-wme}

In current implementations of \hyperref[acronyms_WME]{WME}, the \hyperref[acronyms_WME]{WME} estimate of the causal effect (\(\hat{\beta}_{WME}\)) is calculated as described in Bowden et al\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}, and the 95\% \hyperref[acronyms_CI]{CI} is generated separately using bootstrapping, though notably not using the method described above.

The bootstrapping process begins similarly, with re-sampling undertaken (a default of \(k = 1000\) times) to generate \(k\) bootstrap samples \(x^{*1}, x^{*2},...x^{*k}\). Each individual bootstrap sample \(x^{*i}\) is used to estimate the causal effect using the \hyperref[acronyms_WME]{WME} method \(\hat{\beta}^{*i}_{WME}\), and thus a bootstrap distribution of \(k\) values of \hyperref[acronyms_WME]{WME} is created, \(\hat{\beta}^{*1}_{WME},  \hat{\beta}^{*2}_{WME}....\hat{\beta}^{*k}_{WME}\).

At this stage, however, the bootstrap distribution is then assumed to be approximately normally distributed without verifying this assumption. The 95\% \hyperref[acronyms_CI]{CI} of the bootstrap estimate is then calculated as 1.96 \hyperref[acronyms_SD]{SD}s of the bootstrap distribution either side of the mean estimate, i.e.~\(\hat{\beta}_{WME} \pm 1.96  \times SE\). This approach may be problematic for several reasons.

Although \hyperref[acronyms_CLT]{CLT} leads us to expect that the bootstrap distribution will approach normality as the number of bootstrap iterations \(k\) increases, the extent to which this occurs for a given \(k\) may depend on the inital distribution of values in the population \(X\), and so also on the distribution in the sample/bootstrap population \(x\). If the true distribution of values is very non-normal, as may be the case for traits determined by complex genetic and environmental influences, it may take relatively more bootstrap iterations for the bootstrap distribution to become sufficiently normal to assume mean and \hyperref[acronyms_SD]{SD} accurately describe it.

Additionally, the bootstrap \hyperref[acronyms_SE]{SE} is inversely proportional to the number of bootstrap iterations \(k\), as opposed to the usual standard error (given by \(SE = \frac{SD}{\sqrt{n}}\)), which is inversely proportional to the square root of the sample size \(n\). It is therefore possible to generate smaller \hyperref[acronyms_SE]{SE}s by increasing the number of bootstrap samples obtained. This may lead to false confidence in estimates generated despite potential issues with initial sample \(x\), e.g.~if it too small, or sampled in such a way that it is not representative of the underlying population \(X\). Although such issues are inherent to any bootstrapping approaches, the usual method of generating bootstrapped \hyperref[acronyms_CI]{CI}s detailed above uses more information (i.e.~using the entire bootstrap distribution) to generate these values than the parameter-based \(estimate \pm 1.96 \times SE\) method (i.e.~using approximate summary statistics to represent the distribution). The usual method of bootstrap \hyperref[acronyms_CI]{CI} generation may therefore be expected to highlight any variation or uncertainty present more readily than the parameter-based approach; this would be represented as wider \hyperref[acronyms_CI]{CI}s.

\newpage

\section{Appendix: Simulation Code}\label{appendix-sim}

\subsection{Generating Data and Models}\label{appendix-sim-gen}

The data generating model used was from Appendix 3 of Bowden et al\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}; the relevant section describing their model is reproduced below:

\begin{quote}
\emph{``\ldots{}}
\end{quote}

\begin{quote}
\begin{equation} 
U_i = \sum^J_{j=1} \phi_jG_{ij} + \epsilon_i^U
\end{equation}
\end{quote}

\begin{quote}
\begin{equation} 
X_i = \sum^J_{j=1} \gamma_jG_{ij} + U_i + \epsilon_i^X
\end{equation}
\end{quote}

\begin{quote}
\begin{equation} 
Y_i = \sum^J_{j=1} \alpha_jG_{ij} + \beta X_i + U_i + \epsilon_i^Y
\end{equation}
\end{quote}

\begin{quote}
\emph{for participants indexed by \(i = 1, . . . , N\), and genetic instruments indexed by \(j = 1, . . . , J\).}
\end{quote}

\begin{quote}
\emph{The error terms \(\epsilon_i^U , \epsilon_i^X\) and \(\epsilon_i^Y\) were each drawn independently from standard normal distributions. The genetic effects on the exposure γj are drawn from a uniform distribution between 0.03 and 0.1. Pleiotropic effects \(\alpha_j\) and \(\phi_j\) were set to zero if the genetic instrument was a valid instrumental variable. Otherwise (with probability 0.1, 0.2, or 0.3):}
\end{quote}

\begin{quote}
\emph{1. In Scenario 1 (balanced pleiotropy, InSIDE satisfied), the \(\alpha_j\) parameter was drawn from a uniform distribution between −0.2 and 0.2.}
\end{quote}

\begin{quote}
\emph{2. In Scenario 2 (directional pleiotropy, InSIDE satisfied), the \(\alpha_j\) parameter was drawn from a uniform distribution between 0 and 0.2.}
\end{quote}

\begin{quote}
\emph{3. In Scenario 3 (directional pleiotropy, InSIDE not satisfied), the \(\phi_j\) parameter was drawn from a uniform distribution between −0.2 and 0.2.}
\end{quote}

\begin{quote}
\emph{The causal effect of the exposure on the outcome was either \(\beta X = 0\) (null causal effect) or \(\beta X = 0.1\) (positive causal effect). A total of 10 000 simulated datasets were generated for sample sizes of N = 10 000 and 20 {[}sic{]} participants. Only the summary data, that is genetic associations with the exposure and with the outcome and their standard errors as estimated by univariate regression on the genetic instruments in turn, were used by the analysis methods. In the two-sample setting, data were generated on 2N participants, and genetic associations with the exposure were estimated in the first N participants, and genetic associations with the outcome in the second N participants.''}\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}
\end{quote}

To reproduce this model, code was written in R to generate the relevant participant level data. First, a function (\texttt{get\_simulated\_MR\_data}) was written which included parameters specified by Bowden et al, and also to allow testing of data simulation:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Define function to create data generating model}
\CommentTok{\# Arguments/default values based on Bowden et al}
\NormalTok{get\_simulated\_MR\_data }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(}\AttributeTok{n\_participants =} \FunctionTok{as.integer}\NormalTok{(), }
                                  \AttributeTok{n\_instruments =} \FunctionTok{as.integer}\NormalTok{(),}
                                  \AttributeTok{n\_datasets =} \FunctionTok{as.integer}\NormalTok{(),}
                                  \AttributeTok{prop\_invalid =} \FloatTok{0.1}\NormalTok{,}
                                  \AttributeTok{causal\_effect =} \ConstantTok{TRUE}\NormalTok{,}
                                  \AttributeTok{balanced\_pleio =} \ConstantTok{TRUE}\NormalTok{,}
                                  \AttributeTok{InSIDE\_satisfied =} \ConstantTok{TRUE}\NormalTok{,}
                                  \AttributeTok{rand\_error =} \ConstantTok{TRUE}\NormalTok{,      }\CommentTok{\# remove random errors, for testing}
                                  \AttributeTok{two\_sample =} \ConstantTok{TRUE}\NormalTok{,      }\CommentTok{\# 1{-} or 2{-}sample MR toggle, for testing}
                                  \AttributeTok{beta\_val =} \FloatTok{0.1}\NormalTok{,         }\CommentTok{\# size of causal effect}
                                  \AttributeTok{allele\_freq\_min =} \FloatTok{0.4}\NormalTok{,  }\CommentTok{\# frequency of effect allele 0.01/0.99}
                                  \AttributeTok{allele\_freq\_max =} \FloatTok{0.6}\NormalTok{,  }\CommentTok{\#?0.4/0.6}
                                  \AttributeTok{gamma\_min =} \FloatTok{0.03}\NormalTok{,       }\CommentTok{\# size of pleiotropic effects on exposure}
                                  \AttributeTok{gamma\_max =} \FloatTok{0.1}\NormalTok{,}
                                  \AttributeTok{alpha\_min =} \SpecialCharTok{{-}}\FloatTok{0.2}\NormalTok{,       }\CommentTok{\# size of pleiotropic effects on outcome}
                                  \AttributeTok{alpha\_max =} \FloatTok{0.2}\NormalTok{,}
                                  \AttributeTok{phi\_min =} \SpecialCharTok{{-}}\FloatTok{0.2}\NormalTok{,         }\CommentTok{\# size of additional pleiotropic effects}
                                  \AttributeTok{phi\_max =} \FloatTok{0.2}\NormalTok{,          }\CommentTok{\# when InSIDE not satisfied}
                                  \AttributeTok{seed =} \DecValTok{14101583}\NormalTok{)\{       }\CommentTok{\# Set seed for reproducibility}
  
  \CommentTok{\# Set seed to ensure comparability across scenarios}
  \FunctionTok{set.seed}\NormalTok{(seed)}
  
  \CommentTok{\# Initialise blank lists to receive datasets for}
  \CommentTok{\# each of:}
  \CommentTok{\#     U (vector: unmeasured confounding exposures per participant), }
  \CommentTok{\#     X (vector: exposure:outcome associations estimated per participant) }
  \CommentTok{\#     Y (vector: gene:outcome association estimated per participant), }
  \CommentTok{\#     G (Matrices: Genotype data)}
  \CommentTok{\#}
  \CommentTok{\#     gamma (vector: pleiotropic effects of each instrument on exposure)}
  \CommentTok{\#     alpha (vector: pleiotropic effects of each instrument on outcome)}
  \CommentTok{\#     phi (vector: additional pleiotropic effects of each instrument when InSIDE }
  \CommentTok{\#     assumption not satisfied)}
\NormalTok{  U\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\NormalTok{  X\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\NormalTok{  Y\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\NormalTok{  G\_X\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\NormalTok{  G\_Y\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
  
\NormalTok{  gamma\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\NormalTok{  alpha\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\NormalTok{  phi\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}

  
\NormalTok{  n\_participants\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\NormalTok{  n\_instruments\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\NormalTok{  prop\_invalid\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\NormalTok{  beta\_val\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}

  
  
  \CommentTok{\# {-}{-}{-} Assign features common to all datasets {-}{-}{-} \#}
  
  \CommentTok{\# size of causal effect}
\NormalTok{  beta }\OtherTok{\textless{}{-}} \FunctionTok{if\_else}\NormalTok{(causal\_effect }\SpecialCharTok{==} \ConstantTok{TRUE}\NormalTok{, }
\NormalTok{                  beta\_val,}
                  \DecValTok{0}\NormalTok{)}
  
  \CommentTok{\# create vector of participant indices for 1st n participants}
  \CommentTok{\# i.e. participants used for estimating gene:exposure coefficient}
\NormalTok{  sample\_1\_ref }\OtherTok{\textless{}{-}} \DecValTok{1}\SpecialCharTok{:}\NormalTok{n\_participants        }
  
  
  \CommentTok{\# Default is to estimate gene:outcome coefficient from different sample}
  \CommentTok{\# to gene:exposure coefficient (i.e. simulating 2{-}sample MR)}
  \CommentTok{\# two\_sample == FALSE toggles to single sample for testing simulation}
  \FunctionTok{ifelse}\NormalTok{(two\_sample }\SpecialCharTok{==} \ConstantTok{FALSE}\NormalTok{,}
\NormalTok{         sample\_2\_ref }\OtherTok{\textless{}{-}}\NormalTok{ sample\_1\_ref, }\CommentTok{\# 1 sample MR}
\NormalTok{         sample\_2\_ref }\OtherTok{\textless{}{-}}\NormalTok{ (n\_participants}\SpecialCharTok{+}\DecValTok{1}\NormalTok{)}\SpecialCharTok{:}\NormalTok{(}\DecValTok{2}\SpecialCharTok{*}\NormalTok{n\_participants)) }\CommentTok{\# 2 sample MR}
  
  \CommentTok{\# {-}{-}{-} Set characteristics for each genetic instrument {-}{-}{-} \# }
  
  \CommentTok{\# Set genetic effects of each instrument on the exposure,}
  \CommentTok{\# drawn from uniform distribution, min/max as per Bowden }
  \CommentTok{\# et al}
\NormalTok{  gamma\_vect }\OtherTok{\textless{}{-}} \FunctionTok{runif}\NormalTok{(}\AttributeTok{n =}\NormalTok{ n\_instruments,}
                      \AttributeTok{min =}\NormalTok{ gamma\_min,}
                      \AttributeTok{max =}\NormalTok{ gamma\_max)}
  
  
  
  \CommentTok{\# Set which instruments invalid, 0 = valid, 1 = invalid}
\NormalTok{  invalid\_instrument\_vect }\OtherTok{\textless{}{-}} \FunctionTok{rbinom}\NormalTok{(}\AttributeTok{n =}\NormalTok{ n\_instruments,}
                                    \AttributeTok{size =} \DecValTok{1}\NormalTok{, }
                                    \AttributeTok{prob =}\NormalTok{ prop\_invalid)}
  
  
  \CommentTok{\# Probability of effect allele set per dataset  }
  \CommentTok{\# for each instrument, default value set at  }
  \CommentTok{\# random between 0.01{-}0.99 (i.e. both effect +}
  \CommentTok{\# reference are common alleles)}
\NormalTok{  allele\_freq\_vect }\OtherTok{\textless{}{-}} \FunctionTok{runif}\NormalTok{(}\AttributeTok{n =}\NormalTok{ n\_instruments,}
                            \AttributeTok{min =}\NormalTok{ allele\_freq\_min,}
                            \AttributeTok{max =}\NormalTok{ allele\_freq\_max)}
  
  \CommentTok{\# Set pleiotropic effects on outcome, Scenarios and }
  \CommentTok{\# min/max from Bowden et al}
\NormalTok{  alpha\_vect }\OtherTok{\textless{}{-}} \FunctionTok{double}\NormalTok{() }\CommentTok{\# Pleiotropic effects of instruments on outcome}
\NormalTok{  phi\_vect }\OtherTok{\textless{}{-}} \FunctionTok{double}\NormalTok{() }\CommentTok{\# Pleiotropic effects of confounders on outcome}
  
  
  \ControlFlowTok{for}\NormalTok{(j }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\NormalTok{n\_instruments)\{}
    \FunctionTok{ifelse}\NormalTok{(invalid\_instrument\_vect[j] }\SpecialCharTok{==} \DecValTok{0}\NormalTok{, }\CommentTok{\# alpha = 0 if valid}
\NormalTok{           alpha\_vect[j] }\OtherTok{\textless{}{-}} \DecValTok{0}\NormalTok{,}
           \FunctionTok{ifelse}\NormalTok{(balanced\_pleio }\SpecialCharTok{==} \ConstantTok{TRUE}\NormalTok{,}
\NormalTok{                  alpha\_vect[j] }\OtherTok{\textless{}{-}} \FunctionTok{runif}\NormalTok{(}\AttributeTok{n =} \DecValTok{1}\NormalTok{, }\CommentTok{\# balanced}
                                         \AttributeTok{min =}\NormalTok{ alpha\_min,}
                                         \AttributeTok{max =}\NormalTok{ alpha\_max),}
\NormalTok{                  alpha\_vect[j] }\OtherTok{\textless{}{-}} \FunctionTok{runif}\NormalTok{(}\AttributeTok{n =} \DecValTok{1}\NormalTok{, }\CommentTok{\# directional}
                                         \AttributeTok{min =} \DecValTok{0}\NormalTok{,}
                                         \AttributeTok{max =}\NormalTok{ alpha\_max)}
\NormalTok{           )}
\NormalTok{    )}
    

    \CommentTok{\# Assign default phi = 0 unless directional pleiotropy \& }
    \CommentTok{\# InSIDE assumption not satisfied \& genetic instrument invalid}
    \ControlFlowTok{if}\NormalTok{(balanced\_pleio }\SpecialCharTok{==} \ConstantTok{FALSE} \SpecialCharTok{\&}\NormalTok{ InSIDE\_satisfied }\SpecialCharTok{==} \ConstantTok{FALSE}\NormalTok{)\{}
      \FunctionTok{ifelse}\NormalTok{(invalid\_instrument\_vect[j] }\SpecialCharTok{==} \DecValTok{0}\NormalTok{,}
\NormalTok{             phi\_vect[j] }\OtherTok{\textless{}{-}} \DecValTok{0}\NormalTok{,}
\NormalTok{             phi\_vect[j] }\OtherTok{\textless{}{-}} \FunctionTok{runif}\NormalTok{(}\AttributeTok{n =} \DecValTok{1}\NormalTok{,}
                                  \AttributeTok{min =}\NormalTok{ phi\_min,}
                                  \AttributeTok{max =}\NormalTok{ phi\_max)}
\NormalTok{      )}
      
\NormalTok{    \}}
    \ControlFlowTok{else}\NormalTok{\{}
\NormalTok{      phi\_vect[j] }\OtherTok{\textless{}{-}} \DecValTok{0}
\NormalTok{    \}}
\NormalTok{  \}}
  
  \CommentTok{\# Re{-}set seed to ensure consistency across datasets}
  \CommentTok{\# N.B. above two if/ifelse statements cause de{-}sync }
  \CommentTok{\# of number of randomised functions between valid/invalid}
  \FunctionTok{set.seed}\NormalTok{(seed)}
  
  \CommentTok{\# {-}{-}{-} Create separate datasets {-}{-}{-} \#}
  
  \CommentTok{\# Create N datasets by simulating genotype matrices with}
  \CommentTok{\# 1 row per participant, 1 column per genetic instrument}
  \CommentTok{\# Use these to estimate U, X + Y}
  
  \ControlFlowTok{for}\NormalTok{(n }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\NormalTok{n\_datasets)\{}
    
    \CommentTok{\# {-}{-}{-} Create matrix of genotypes {-}{-}{-} \#}

    \CommentTok{\# Assign genotypes by sampling from binomial distribution}
    \CommentTok{\# twice (as two alleles) per participant with probability}
    \CommentTok{\# equal to frequency of effect allele}
    \CommentTok{\# Create twice as many genotypes as participants in sample}
    \CommentTok{\# to simulate 2 sample MR, i.e. first half used to estimate}
    \CommentTok{\# Gene:Exposure, second half used to estimate Gene:Outcome}
    
    \CommentTok{\# Matrix where columns are instruments, rows are participants}
    \CommentTok{\# Values 0, 1 or 2}
    \CommentTok{\# 0 = reference, i.e. zero effect alleles, }
    \CommentTok{\# 1 = 1 effect allele, 2 = 2 effect alleles }
    
\NormalTok{    G\_mat }\OtherTok{\textless{}{-}} \FunctionTok{matrix}\NormalTok{(}\FunctionTok{rbinom}\NormalTok{(}\AttributeTok{n =} \DecValTok{2} \SpecialCharTok{*}\NormalTok{ n\_participants }\SpecialCharTok{*}\NormalTok{ n\_instruments,}
                           \AttributeTok{size =} \DecValTok{2}\NormalTok{,}
                           \AttributeTok{prob =} \FunctionTok{rep}\NormalTok{(allele\_freq\_vect, }\DecValTok{2} \SpecialCharTok{*}\NormalTok{ n\_participants)),}
                    \AttributeTok{nrow =} \DecValTok{2} \SpecialCharTok{*}\NormalTok{ n\_participants,}
                    \AttributeTok{ncol =}\NormalTok{ n\_instruments,}
                    \AttributeTok{byrow =} \ConstantTok{TRUE}\NormalTok{)}
    
    
    \CommentTok{\# Create error terms for U, X + Y per participant,}
    \CommentTok{\# each drawn from standard normal distribution}
    \CommentTok{\# unless random error turned off (for testing)}
    
    \FunctionTok{ifelse}\NormalTok{(rand\_error }\SpecialCharTok{==} \ConstantTok{TRUE}\NormalTok{,}
\NormalTok{           U\_epsilon\_vect }\OtherTok{\textless{}{-}} \FunctionTok{rnorm}\NormalTok{(}\AttributeTok{n =} \DecValTok{2} \SpecialCharTok{*}\NormalTok{ n\_participants),}
\NormalTok{           U\_epsilon\_vect }\OtherTok{\textless{}{-}} \FunctionTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{2} \SpecialCharTok{*}\NormalTok{ n\_participants))}
    
    \FunctionTok{ifelse}\NormalTok{(rand\_error }\SpecialCharTok{==} \ConstantTok{TRUE}\NormalTok{,}
\NormalTok{           X\_epsilon\_vect }\OtherTok{\textless{}{-}} \FunctionTok{rnorm}\NormalTok{(}\AttributeTok{n =}\NormalTok{ n\_participants),}
\NormalTok{           X\_epsilon\_vect }\OtherTok{\textless{}{-}} \FunctionTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{, n\_participants))}
    
    \FunctionTok{ifelse}\NormalTok{(rand\_error }\SpecialCharTok{==} \ConstantTok{TRUE}\NormalTok{,}
\NormalTok{           Y\_epsilon\_vect }\OtherTok{\textless{}{-}} \FunctionTok{rnorm}\NormalTok{(}\AttributeTok{n =}\NormalTok{ n\_participants),}
\NormalTok{           Y\_epsilon\_vect }\OtherTok{\textless{}{-}} \FunctionTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{, n\_participants))}
    
    
    \CommentTok{\# {-}{-}{-} Combine Gene matrix/parameters to recreate model {-}{-}{-} \#}
    
    \CommentTok{\# Create vectors of estimates for U, X and Y per individual,}
    \CommentTok{\# i.e. Ui, Xi and Yi. Uses matrix inner product operator " \%*\%" }
    \CommentTok{\# https://stackoverflow.com/questions/22060515/the{-}r{-}operator }
    \CommentTok{\# http://matrixmultiplication.xyz/}

    \CommentTok{\#     U (vector: unmeasured confounding exposures per participant), }
    \CommentTok{\#     X (vector: exposure:outcome associations estimated per participant) }
    \CommentTok{\#     Y (vector: gene:outcome association estimated per participant) }
    
\NormalTok{    Ui\_vect }\OtherTok{\textless{}{-}}\NormalTok{  G\_mat }\SpecialCharTok{\%*\%}\NormalTok{ phi\_vect }\SpecialCharTok{+}\NormalTok{ U\_epsilon\_vect}
    
\NormalTok{    Xi\_vect }\OtherTok{\textless{}{-}}\NormalTok{  G\_mat[sample\_1\_ref, ] }\SpecialCharTok{\%*\%}\NormalTok{ gamma\_vect }\SpecialCharTok{+} 
\NormalTok{      Ui\_vect[sample\_1\_ref, ] }\SpecialCharTok{+} 
\NormalTok{      X\_epsilon\_vect}
    
\NormalTok{    Yi\_vect }\OtherTok{\textless{}{-}}\NormalTok{  G\_mat[sample\_2\_ref, ] }\SpecialCharTok{\%*\%}\NormalTok{ alpha\_vect }\SpecialCharTok{+} 
\NormalTok{      beta }\SpecialCharTok{*}\NormalTok{ Xi\_vect }\SpecialCharTok{+} 
\NormalTok{      Ui\_vect[sample\_2\_ref, ] }\SpecialCharTok{+} 
\NormalTok{      Y\_epsilon\_vect}
    
    
    \CommentTok{\# Add vectors of estimates from this dataset to lists of }
    \CommentTok{\# estimates from all datasets}
\NormalTok{    U\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ Ui\_vect}
    
\NormalTok{    X\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ Xi\_vect}
    
\NormalTok{    Y\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ Yi\_vect}
    
\NormalTok{    G\_X\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ G\_mat[sample\_1\_ref, ]}
    
\NormalTok{    G\_Y\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ G\_mat[sample\_2\_ref, ]}
    
    
    \CommentTok{\# Include actual parameter values generated for simulation }
\NormalTok{    alpha\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ alpha\_vect}
    
\NormalTok{    gamma\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ gamma\_vect}
    
\NormalTok{    phi\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ phi\_vect}
    
    \CommentTok{\# Include inputs for reference/testing }
\NormalTok{    n\_participants\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ n\_participants}
\NormalTok{    n\_instruments\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ n\_instruments}
\NormalTok{    prop\_invalid\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ prop\_invalid}
\NormalTok{    beta\_val\_list[[n]] }\OtherTok{\textless{}{-}}\NormalTok{ beta\_val}
   
     
\NormalTok{  \}}
  
  
  
  \CommentTok{\# {-}{-}{-} Combine all outputs to return {-}{-}{-} \#}
  
\NormalTok{  combined\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{(}\AttributeTok{U =}\NormalTok{ U\_list,         }\CommentTok{\# Estimates }
                        \AttributeTok{X =}\NormalTok{ X\_list, }
                        \AttributeTok{Y =}\NormalTok{ Y\_list,}
                        
                        \AttributeTok{G\_X =}\NormalTok{ G\_X\_list,     }\CommentTok{\# Genotypes of 1st sample}
                        \AttributeTok{G\_Y =}\NormalTok{ G\_Y\_list,     }\CommentTok{\# Genotypes of 2nd sample}
                        
                        \AttributeTok{alpha =}\NormalTok{ alpha\_list, }\CommentTok{\# Actual values for validating simulation}
                        \AttributeTok{gamma =}\NormalTok{ gamma\_list,}
                        \AttributeTok{phi =}\NormalTok{ phi\_list,}
                        
                        \AttributeTok{n\_participants =}\NormalTok{ n\_participants\_list, }\CommentTok{\# Inputs}
                        \AttributeTok{n\_instruments =}\NormalTok{ n\_instruments\_list,}
                        \AttributeTok{prop\_invalid =}\NormalTok{ prop\_invalid\_list,}
                        \AttributeTok{beta\_val =}\NormalTok{ beta\_val\_list}
\NormalTok{  )}
  
  \FunctionTok{return}\NormalTok{(combined\_list)}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

This initial simulation function generated data in the following format:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Check data produced in expected format}
\CommentTok{\#set.seed(1701)}
\NormalTok{test\_data\_sim }\OtherTok{\textless{}{-}} \FunctionTok{get\_simulated\_MR\_data}\NormalTok{(}\AttributeTok{n\_participants =} \DecValTok{1000}\NormalTok{,}
                                       \AttributeTok{n\_instruments =} \DecValTok{25}\NormalTok{,}
                                       \AttributeTok{n\_datasets =} \DecValTok{2}\NormalTok{,}
                                       \AttributeTok{prop\_invalid =} \FloatTok{0.3}\NormalTok{,}
                                       \AttributeTok{rand\_error =} \ConstantTok{FALSE}\NormalTok{,}
                                       \AttributeTok{causal\_effect =} \ConstantTok{TRUE}\NormalTok{,}
                                       \AttributeTok{balanced\_pleio =} \ConstantTok{TRUE}\NormalTok{,}
                                       \AttributeTok{InSIDE\_satisfied =} \ConstantTok{TRUE}\NormalTok{)}

\FunctionTok{str}\NormalTok{(test\_data\_sim)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## List of 12
##  $ U             :List of 2
##   ..$ : num [1:2000, 1] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ : num [1:2000, 1] 0 0 0 0 0 0 0 0 0 0 ...
##  $ X             :List of 2
##   ..$ : num [1:1000, 1] 1.59 1.29 1.02 1.89 1.49 ...
##   ..$ : num [1:1000, 1] 1.85 1.84 2.18 1.16 1.44 ...
##  $ Y             :List of 2
##   ..$ : num [1:1000, 1] 0.0704 0.1351 0.1589 0.0944 0.0161 ...
##   ..$ : num [1:1000, 1] 0.59017 0.10039 0.00743 0.27896 0.27746 ...
##  $ G_X           :List of 2
##   ..$ : int [1:1000, 1:25] 2 0 1 2 1 0 1 0 0 2 ...
##   ..$ : int [1:1000, 1:25] 0 1 2 1 1 1 0 2 0 2 ...
##  $ G_Y           :List of 2
##   ..$ : int [1:1000, 1:25] 1 1 0 1 2 0 1 0 1 2 ...
##   ..$ : int [1:1000, 1:25] 1 1 1 1 0 1 2 0 2 0 ...
##  $ alpha         :List of 2
##   ..$ : num [1:25] 0 0 0.1157 0 -0.0634 ...
##   ..$ : num [1:25] 0 0 0.1157 0 -0.0634 ...
##  $ gamma         :List of 2
##   ..$ : num [1:25] 0.0938 0.0808 0.0755 0.0342 0.0443 ...
##   ..$ : num [1:25] 0.0938 0.0808 0.0755 0.0342 0.0443 ...
##  $ phi           :List of 2
##   ..$ : num [1:25] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ : num [1:25] 0 0 0 0 0 0 0 0 0 0 ...
##  $ n_participants:List of 2
##   ..$ : num 1000
##   ..$ : num 1000
##  $ n_instruments :List of 2
##   ..$ : num 25
##   ..$ : num 25
##  $ prop_invalid  :List of 2
##   ..$ : num 0.3
##   ..$ : num 0.3
##  $ beta_val      :List of 2
##   ..$ : num 0.1
##   ..$ : num 0.1
\end{verbatim}

A function (\texttt{get\_models}) was then written to create linear models from each dataset generated as per Bowden et al:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Create plotting tibble with Mean/SD X + Y grouped by}
\CommentTok{\# Dataset + instrument}
\NormalTok{get\_models }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(sim)\{}
    
\NormalTok{  output\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
  
  \CommentTok{\# Create linear models per dataset to get coefficients}
  \CommentTok{\# for gene:exposure association (coeff\_G\_X) and gene:outcome}
  \CommentTok{\# association (coeff\_G\_Y)}
  \ControlFlowTok{for}\NormalTok{(dataset }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\FunctionTok{length}\NormalTok{(sim}\SpecialCharTok{$}\NormalTok{X))\{}
    
\NormalTok{    X }\OtherTok{\textless{}{-}}\NormalTok{ sim}\SpecialCharTok{$}\NormalTok{X[[dataset]]}
\NormalTok{    Y }\OtherTok{\textless{}{-}}\NormalTok{ sim}\SpecialCharTok{$}\NormalTok{Y[[dataset]]}
\NormalTok{    Instruments\_X }\OtherTok{\textless{}{-}}\NormalTok{ sim}\SpecialCharTok{$}\NormalTok{G\_X[[dataset]]}
\NormalTok{    Instruments\_Y }\OtherTok{\textless{}{-}}\NormalTok{ sim}\SpecialCharTok{$}\NormalTok{G\_Y[[dataset]]}
    
\NormalTok{    alpha }\OtherTok{\textless{}{-}}\NormalTok{ sim}\SpecialCharTok{$}\NormalTok{alpha[[dataset]]}
\NormalTok{    gamma }\OtherTok{\textless{}{-}}\NormalTok{ sim}\SpecialCharTok{$}\NormalTok{gamma[[dataset]]}
\NormalTok{    phi }\OtherTok{\textless{}{-}}\NormalTok{ sim}\SpecialCharTok{$}\NormalTok{phi[[dataset]]}
\NormalTok{    beta }\OtherTok{\textless{}{-}}\NormalTok{ sim}\SpecialCharTok{$}\NormalTok{beta\_val[[dataset]]}
\NormalTok{    prop\_invalid }\OtherTok{\textless{}{-}}\NormalTok{ sim}\SpecialCharTok{$}\NormalTok{prop\_invalid[[dataset]]}
\NormalTok{    n\_instruments }\OtherTok{\textless{}{-}}\NormalTok{ sim}\SpecialCharTok{$}\NormalTok{n\_instruments[[dataset]]}
\NormalTok{    n\_participants}\OtherTok{\textless{}{-}}\NormalTok{ sim}\SpecialCharTok{$}\NormalTok{n\_participants[[dataset]]}
    
    
    \CommentTok{\# Model for gene:exposure ***0 + deleted from X\_lm and Y\_lm***}
\NormalTok{    X\_lm }\OtherTok{\textless{}{-}} \FunctionTok{lm}\NormalTok{(X }\SpecialCharTok{\textasciitilde{}} \DecValTok{0} \SpecialCharTok{+}\NormalTok{ Instruments\_X)}
\NormalTok{    coeff\_G\_X\_vect }\OtherTok{\textless{}{-}} \FunctionTok{coef}\NormalTok{(}\FunctionTok{summary}\NormalTok{(X\_lm))[}\DecValTok{1}\SpecialCharTok{:}\NormalTok{(}\FunctionTok{ncol}\NormalTok{(Instruments\_X)), }\DecValTok{1}\NormalTok{]}
\NormalTok{    SE\_coeff\_G\_X\_vect }\OtherTok{\textless{}{-}} \FunctionTok{coef}\NormalTok{(}\FunctionTok{summary}\NormalTok{(X\_lm))[}\DecValTok{1}\SpecialCharTok{:}\NormalTok{(}\FunctionTok{ncol}\NormalTok{(Instruments\_X)), }\DecValTok{2}\NormalTok{]}
    
\NormalTok{    R2\_stat }\OtherTok{\textless{}{-}} \FunctionTok{summary}\NormalTok{(}\FunctionTok{lm}\NormalTok{(X }\SpecialCharTok{\textasciitilde{}}\NormalTok{ Instruments\_X))}\SpecialCharTok{$}\NormalTok{r.squared}
\NormalTok{    F\_stat }\OtherTok{\textless{}{-}} \FunctionTok{summary}\NormalTok{(}\FunctionTok{lm}\NormalTok{(X }\SpecialCharTok{\textasciitilde{}}\NormalTok{ Instruments\_X))}\SpecialCharTok{$}\NormalTok{fstatistic[[}\DecValTok{1}\NormalTok{]]}

    
    \CommentTok{\# Model for gene:outcome}
\NormalTok{    Y\_lm }\OtherTok{\textless{}{-}} \FunctionTok{lm}\NormalTok{(Y }\SpecialCharTok{\textasciitilde{}} \DecValTok{0} \SpecialCharTok{+}\NormalTok{ Instruments\_Y)}
\NormalTok{    coeff\_G\_Y\_vect }\OtherTok{\textless{}{-}} \FunctionTok{coef}\NormalTok{(}\FunctionTok{summary}\NormalTok{(Y\_lm))[}\DecValTok{1}\SpecialCharTok{:}\NormalTok{(}\FunctionTok{ncol}\NormalTok{(Instruments\_Y)), }\DecValTok{1}\NormalTok{] }
\NormalTok{    SE\_coeff\_G\_Y\_vect }\OtherTok{\textless{}{-}} \FunctionTok{coef}\NormalTok{(}\FunctionTok{summary}\NormalTok{(Y\_lm))[}\DecValTok{1}\SpecialCharTok{:}\NormalTok{(}\FunctionTok{ncol}\NormalTok{(Instruments\_Y)), }\DecValTok{2}\NormalTok{]}
    
\NormalTok{    output\_list[[dataset]] }\OtherTok{\textless{}{-}} \FunctionTok{as\_tibble}\NormalTok{(}\FunctionTok{list}\NormalTok{(}\AttributeTok{dataset =}\NormalTok{ dataset,}
                                             \AttributeTok{Instrument =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\SpecialCharTok{:}\FunctionTok{ncol}\NormalTok{(Instruments\_X)),}
                                             \AttributeTok{coeff\_G\_X =}\NormalTok{ coeff\_G\_X\_vect,}
                                             \AttributeTok{coeff\_G\_X\_SE =}\NormalTok{ SE\_coeff\_G\_X\_vect,}
                                             \AttributeTok{gamma =}\NormalTok{ gamma,}
                                             \AttributeTok{F\_stat =}\NormalTok{ F\_stat,}
                                             \AttributeTok{R2\_stat =}\NormalTok{ R2\_stat,}
                                             \AttributeTok{coeff\_G\_Y =}\NormalTok{ coeff\_G\_Y\_vect,}
                                             \AttributeTok{coeff\_G\_Y\_SE =}\NormalTok{ SE\_coeff\_G\_Y\_vect,}
                                             \AttributeTok{alpha =}\NormalTok{ alpha,}
                                             \AttributeTok{phi =}\NormalTok{ phi,}
                                             \AttributeTok{beta =}\NormalTok{ beta,}
                                             \AttributeTok{prop\_invalid =}\NormalTok{ prop\_invalid,}
                                             \AttributeTok{n\_instruments =}\NormalTok{ n\_instruments,}
                                             \AttributeTok{n\_participants =}\NormalTok{ n\_participants),}
                                        \AttributeTok{.name\_repair =} \StringTok{"unique"}\NormalTok{)}
\NormalTok{  \}}
  
  \FunctionTok{return}\NormalTok{(output\_list)}
  
\NormalTok{\}  }
\end{Highlighting}
\end{Shaded}

These models generated estimates of the coefficient of gene-exposure association (\texttt{coeff\_G\_X}), coefficient of gene-outcome association (\texttt{coeff\_G\_Y}), and the relevant standard errors of these estimates. The values of parameters inputted were also returned to aid in further testing of data/model generation, i.e.~actual gene-exposure associations (\texttt{gamma}), pleiotropic effects of invalid instruments (\texttt{alpha}), additional pleiotropic effects when \hyperref[acronyms_InSIDE]{InSIDE} assumption not satified (\texttt{phi}), causal effect of exposure on outcome (\texttt{beta}) and the proportion of invalid genetic instruments with pleiotropic effects on the outcome (\texttt{prop\_invalid}).

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{test\_extract\_model }\OtherTok{\textless{}{-}} \FunctionTok{get\_models}\NormalTok{(test\_data\_sim)}

\FunctionTok{summary}\NormalTok{(test\_extract\_model[[}\DecValTok{1}\NormalTok{]])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##     dataset    Instrument   coeff_G_X        coeff_G_X_SE      
##  Min.   :1   Min.   : 1   Min.   :0.03419   Min.   :6.659e-17  
##  1st Qu.:1   1st Qu.: 7   1st Qu.:0.05645   1st Qu.:6.807e-17  
##  Median :1   Median :13   Median :0.06823   Median :6.873e-17  
##  Mean   :1   Mean   :13   Mean   :0.06791   Mean   :6.889e-17  
##  3rd Qu.:1   3rd Qu.:19   3rd Qu.:0.08594   3rd Qu.:6.935e-17  
##  Max.   :1   Max.   :25   Max.   :0.09379   Max.   :7.313e-17  
##      gamma             F_stat             R2_stat    coeff_G_Y        
##  Min.   :0.03419   Min.   :6.224e+27   Min.   :1   Min.   :-0.109067  
##  1st Qu.:0.05645   1st Qu.:6.224e+27   1st Qu.:1   1st Qu.: 0.004951  
##  Median :0.06823   Median :6.224e+27   Median :1   Median : 0.006555  
##  Mean   :0.06791   Mean   :6.224e+27   Mean   :1   Mean   : 0.013297  
##  3rd Qu.:0.08594   3rd Qu.:6.224e+27   3rd Qu.:1   3rd Qu.: 0.008890  
##  Max.   :0.09379   Max.   :6.224e+27   Max.   :1   Max.   : 0.162629  
##   coeff_G_Y_SE          alpha                phi         beta      prop_invalid
##  Min.   :0.001550   Min.   :-0.115363   Min.   :0   Min.   :0.1   Min.   :0.3  
##  1st Qu.:0.001579   1st Qu.: 0.000000   1st Qu.:0   1st Qu.:0.1   1st Qu.:0.3  
##  Median :0.001591   Median : 0.000000   Median :0   Median :0.1   Median :0.3  
##  Mean   :0.001598   Mean   : 0.006717   Mean   :0   Mean   :0.1   Mean   :0.3  
##  3rd Qu.:0.001624   3rd Qu.: 0.000000   3rd Qu.:0   3rd Qu.:0.1   3rd Qu.:0.3  
##  Max.   :0.001653   Max.   : 0.156224   Max.   :0   Max.   :0.1   Max.   :0.3  
##  n_instruments n_participants
##  Min.   :25    Min.   :1000  
##  1st Qu.:25    1st Qu.:1000  
##  Median :25    Median :1000  
##  Mean   :25    Mean   :1000  
##  3rd Qu.:25    3rd Qu.:1000  
##  Max.   :25    Max.   :1000
\end{verbatim}

\newpage

\subsection{Testing Generation of Data and Models}\label{appendix-sim-test}

A series of test plots were used to verify that data were simulated as intended under the various conditions specified by input parameters. Test plots were not created for the parameters \texttt{n\_participants}, \texttt{n\_instruments} or \texttt{n\_datasets}, as the functioning of these parameters could be readily inferred from the structure of the datasets outputted, as above.

\subsubsection{Proportion of Invalid Instruments}\label{proportion-of-invalid-instruments}

\leavevmode\newline The \texttt{prop\_invalid} parameter specifies the proportion of invalid genetic instruments simulated, i.e.~the proportion of genetic instruments affecting the outcome via direct/pleiotropic effects, and thus not solely via the exposure of interest. If simulated correctly, increasing the value of \texttt{prop\_invalid} should increase the number of instruments with pleiotropic effects, i.e.~instruments with \texttt{alpha} \(\ne\) 0. With random error terms set to 0 and no causal effect present (i.e.~\texttt{rand\_error\ =\ FALSE} and \texttt{causal\_effect\ =\ FALSE}), the estimated gene-outcome coefficient estimated using any given instrument will equal the pleiotropic effects of that instrument (i.e.~\texttt{coeff\_G\_Y\ =\ alpha}), and therefore will only be non-zero for invalid instruments with non-zero pleiotropic effects on the outcome . Plotting \texttt{coeff\_G\_Y} against \texttt{alpha} for simulated data with no causal effect or random error should therefore yield a graph where

\begin{itemize}
\tightlist
\item
  For valid instruments: gene-outcome coefficient = alpha = 0
\item
  For invalid instruments: gene-outcome coefficient = alpha \(\ne\) 0, with values spread uniformly between \texttt{alpha\_min} and \texttt{alpha\_max}
\end{itemize}

\pandocbounded{\includegraphics[keepaspectratio]{9_Appendices_files/figure-latex/test-plot-prop-invalid-1.pdf}}

\newpage

Similarly, with random error terms set to 0 (\texttt{rand\_error\ =\ FALSE}) and no causal effect present (\texttt{causal\_effect\ =\ FALSE}), gene-exposure coefficients estimated for each instrument should exactly match the actual values simulated, i.e.~\texttt{coeff\_G\_X\ =\ gamma} for all instruments:

\pandocbounded{\includegraphics[keepaspectratio]{9_Appendices_files/figure-latex/test-plot-gamma-1-1.pdf}}

\newpage

\subsubsection{Gene-Exposure Coefficient Versus Gene-Outcome Coefficient Plots}\label{gene-exposure-coefficient-versus-gene-outcome-coefficient-plots}

\leavevmode\newline For the next phase of testing, a function (\texttt{plot\_GY\_GX}) was written to plot the coefficients for gene-exposure versus gene-outcome as estimated using the previously created linear models:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{plot\_GY\_GX }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(model\_tib, }
                       \AttributeTok{plot\_title =} \FunctionTok{as.character}\NormalTok{(}\ConstantTok{NA}\NormalTok{),}
                       \AttributeTok{x\_min =} \DecValTok{0}\NormalTok{,                     }\CommentTok{\# set x{-}axis limits}
                       \AttributeTok{x\_max =} \FloatTok{0.1}\NormalTok{,}
                       \AttributeTok{y\_min =} \SpecialCharTok{{-}}\FloatTok{0.05}\NormalTok{,                 }\CommentTok{\# set x{-}axis limits}
                       \AttributeTok{y\_max =} \FloatTok{0.06}\NormalTok{,}
                       \AttributeTok{beta\_x =} \FloatTok{0.075}\NormalTok{,                }\CommentTok{\# set beta{-}hat position}
                       \AttributeTok{beta\_y =} \FloatTok{0.05}\NormalTok{,}
                       \AttributeTok{hat\_offset =} \FloatTok{0.003}
\NormalTok{)}
\NormalTok{\{}
  
\NormalTok{  model\_tib }\SpecialCharTok{\%\textgreater{}\%} 
    \FunctionTok{mutate}\NormalTok{(}\AttributeTok{Gradient =} \FunctionTok{round}\NormalTok{(}\FunctionTok{coefficients}\NormalTok{(}\FunctionTok{lm}\NormalTok{(coeff\_G\_Y }\SpecialCharTok{\textasciitilde{}} \DecValTok{0} \SpecialCharTok{+}\NormalTok{ coeff\_G\_X)[}\DecValTok{1}\NormalTok{], }\DecValTok{5}\NormalTok{), }
                            \AttributeTok{digits =} \DecValTok{2}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{plot\_template}\NormalTok{() }\SpecialCharTok{+} \CommentTok{\# pre{-}formatted plot template {-} call to ggplot with UoE colours}
    \FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ coeff\_G\_X, }\AttributeTok{y =}\NormalTok{ coeff\_G\_Y) }\SpecialCharTok{+}
    \FunctionTok{geom\_point}\NormalTok{(}\AttributeTok{colour =}\NormalTok{ edin\_bright\_red\_hex, }\AttributeTok{alpha =} \FloatTok{0.3}\NormalTok{) }\SpecialCharTok{+}
    \FunctionTok{geom\_abline}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{intercept =} \DecValTok{0}\NormalTok{, }
                    \AttributeTok{slope =}\NormalTok{ Gradient),}
                \AttributeTok{size =} \DecValTok{1}\NormalTok{,}
                \AttributeTok{colour =}\NormalTok{ edin\_uni\_blue\_hex) }\SpecialCharTok{+}
    \FunctionTok{geom\_text}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{label =} \FunctionTok{paste0}\NormalTok{(}\StringTok{"\textbackslash{}U03B2 = "}\NormalTok{, }\FunctionTok{as.character}\NormalTok{(Gradient))), }\CommentTok{\#beta}
              \AttributeTok{x =}\NormalTok{ beta\_x, }\CommentTok{\# labels with gradient (causal effect estimate)}
              \AttributeTok{y =}\NormalTok{ beta\_y,}
              \AttributeTok{colour =}\NormalTok{ edin\_uni\_blue\_hex, }
              \AttributeTok{hjust =} \DecValTok{0}\NormalTok{, }
              \AttributeTok{data =}\NormalTok{ . }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{slice\_head}\NormalTok{()}\CommentTok{\# prevent over{-}printing}
\NormalTok{    ) }\SpecialCharTok{+}
    \CommentTok{\#label = expression("True" \textasciitilde{} hat(beta)\textasciitilde{} "= 0.25"),}
    \FunctionTok{annotate}\NormalTok{(}\StringTok{"text"}\NormalTok{,}
             \AttributeTok{x =}\NormalTok{ beta\_x,      }\CommentTok{\# add hat to beta}
             \AttributeTok{y =}\NormalTok{ beta\_y }\SpecialCharTok{+}\NormalTok{ hat\_offset,}
             \AttributeTok{label =} \FunctionTok{paste}\NormalTok{(}\StringTok{"\textbackslash{}U02C6"}\NormalTok{),}
             \AttributeTok{colour =}\NormalTok{ edin\_uni\_blue\_hex, }
             \AttributeTok{hjust =} \SpecialCharTok{{-}}\FloatTok{0.4}\NormalTok{,}
             \AttributeTok{vjust =} \FloatTok{0.9}
\NormalTok{    ) }\SpecialCharTok{+}
    \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =}\NormalTok{ plot\_title,}
         \AttributeTok{x =} \StringTok{"Gene{-}Exposure Coefficient"}\NormalTok{,}
         \AttributeTok{y =} \StringTok{"Gene{-}Outcome Coefficient"}\NormalTok{) }\SpecialCharTok{+}
    \FunctionTok{xlim}\NormalTok{(x\_min, x\_max) }\SpecialCharTok{+}
    \FunctionTok{ylim}\NormalTok{(y\_min, y\_max)}
  
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\newpage

With random error terms set to 0 (\texttt{rand\_error\ =\ FALSE}) and no causal effect present, a graph of gene-exposure coefficients versus gene-outcome coefficients should be a straight line through the origin with gradient = 0; causal effect of \(\beta\) = 0.1 present (\texttt{beta\_val\ =\ 0.1}, \texttt{causal\_effect\ =\ TRUE}), the slope of a graph of gene-exposure coefficients versus gene-outcome coefficients from the same sample should be a straight line through the origin with gradient = 0.1:

\pandocbounded{\includegraphics[keepaspectratio]{9_Appendices_files/figure-latex/test-plot-causal-1.pdf}}

\newpage

\subsubsection{Random Errors}\label{random-errors}

\leavevmode\newline Re-plotting the same graphs with non-zero random error terms (\texttt{rand\_error\ =\ TRUE}) should produce similar graphs with Gaussian spread around lines passing through the origin with gradients of 0 and 0.1 for no causal effect and causal effect, respectively:

\pandocbounded{\includegraphics[keepaspectratio]{9_Appendices_files/figure-latex/test-plot-causal-errors-1.pdf}}

\newpage

\subsubsection{One versus Two Sample MR}\label{one-versus-two-sample-mr}

\leavevmode\newline Where gene-exposure coefficients and gene-outcome coefficients are estimated from two separate samples rather than one (i.e.~\texttt{two\_sample\ =\ TRUE}, simulating 2 sample MR), even with random error terms set to zero, error will be introduced into causal effect estimation through random sampling of different combinations of effect alleles. However, where a causal effect is not present, the effect estimated will consistently be zero regardless of the combinations of alleles sampled, so random error should not be introduced:

\pandocbounded{\includegraphics[keepaspectratio]{9_Appendices_files/figure-latex/test-plot-causal-2SMR-1.pdf}}

\newpage

\subsubsection{Invalid Instruments}\label{invalid-instruments}

\leavevmode\newline Where invalid instruments are present (i.e.~\texttt{prop\_invalid} \(\ne\) \texttt{0}) and random error terms are set to 0, graphs of gene-exposure coefficients versus gene-outcome coefficients should be straight lines through the origin and all points representing valid instruments; the invalid instruments should appear as outliers to this line:

\pandocbounded{\includegraphics[keepaspectratio]{9_Appendices_files/figure-latex/test-plot-causal-inval-1.pdf}}

\newpage

\subsubsection{Balanced Versus Directional Pleiotropy}\label{balanced-versus-directional-pleiotropy}

\leavevmode\newline Replotting the above with unbalanced pleiotropy present (\texttt{balanced\_pleio\ =\ FALSE}), the invalid instruments should all appear as outliers in the positive direction, i.e.~steepening the line of best fit and leading to overestimation of the causal effect:

\pandocbounded{\includegraphics[keepaspectratio]{9_Appendices_files/figure-latex/test-plot-causal-unbal-1.pdf}}

\newpage

\subsubsection{InSIDE Assumption and Phi}\label{inside-assumption-and-phi}

\leavevmode\newline The variable phi represents additional pleiotropic effects of each invalid instrument when the \hyperref[acronyms_InSIDE]{InSIDE} assumption is not satisfied. The \hyperref[acronyms_InSIDE]{InSIDE} assumption states that the gene-exposure association is not correlated with the pleiotropic path gene-outcome path of any invalid genetic instruments. This assumption can be violated if e.g.:

\begin{itemize}
\item
  several invalid genetic instruments influence the outcome via the same pleiotropic path
\item
  several invalid genetic instruments are related to the same (unmeasured) confounders of the exposure:outcome relationship, aka correlated pleiotropy.
\end{itemize}

As such, when the \hyperref[acronyms_InSIDE]{InSIDE} assumption is violated, even ``strong'' instruments (i.e.~those with a strong gene-exposure relationship) may not allow accurate estimation of the true causal effect, as pleiotropic effects may scale with instrument strength. If pleiotropic effects are balanced, InSIDE assumption violation may lead to greater imprecision in causal effect estimation; if pleiotropic effects are directional, \hyperref[acronyms_InSIDE]{InSIDE} assumption violation may lead to bias.

Bowden et al\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}} modeled phi as the pleiotropic effects of unmeasured genetic confounders of the exposure:outcome relationship. Phi adds additional error to causal effect estimation in scenarios with directional pleiotropic effects (\texttt{0\ \textless{}\ alpha\ \textless{}\ 0.2}) and \hyperref[acronyms_InSIDE]{InSIDE} assumption violation. As such, switching \texttt{InSIDE\_satisfied} from \texttt{TRUE} to \texttt{FALSE} should add scatter to the linear association expected when plotting alpha versus gene-outcome coefficients with random error terms set to zero:

\pandocbounded{\includegraphics[keepaspectratio]{9_Appendices_files/figure-latex/test-plot-phi-1.pdf}}

\newpage

Setting \texttt{InSIDE\_satisfied\ =\ TRUE} should mean \texttt{phi\ =\ 0}; \texttt{InSIDE\_satisfied=FALSE} should result in \texttt{phi} \(\propto\) gene-outcome coefficient, with scatter only in the positive direction of gene-outcome coefficients given the model also requires directional pleiotropy before \texttt{phi} is used:

\pandocbounded{\includegraphics[keepaspectratio]{9_Appendices_files/figure-latex/test-plot-phi-2-1.pdf}}

\newpage

\subsection{Summary Table}\label{appendix-sim-summ}

A function (\texttt{get\_summary\_MR\_tib\_row}) was written to take models generated from each simulated dataset, estimate causal effect using both weighted median and MR-Hevo methodologies, then output a summary formatted as per Tables 2 \& 3 in Bowden et al\textsuperscript{\citeproc{ref-bowden_consistent_2016}{8}}:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load WME functions}
\FunctionTok{library}\NormalTok{(TwoSampleMR)}

\CommentTok{\# Load RStan {-} needed for MR{-}Hevo}
\FunctionTok{library}\NormalTok{(rstan)}


\CommentTok{\# Run local copy of MR{-}Hevo functions}
\CommentTok{\# Not using full package due to conflicts with Windows}
\FunctionTok{source}\NormalTok{(}\FunctionTok{here}\NormalTok{(}\StringTok{"Script"}\NormalTok{, }\StringTok{"Hevo"}\NormalTok{, }\StringTok{"functions.mrhevo.R"}\NormalTok{))}

\CommentTok{\# Standard set{-}up for RStan}
\FunctionTok{options}\NormalTok{(}\AttributeTok{mc.cores =}\NormalTok{ parallel}\SpecialCharTok{::}\FunctionTok{detectCores}\NormalTok{())}
\FunctionTok{rstan\_options}\NormalTok{(}\AttributeTok{auto\_write =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{save\_dso =} \ConstantTok{TRUE}\NormalTok{)}


\CommentTok{\# Compile model for MR{-}Hevo}
\NormalTok{mr.stanmodel }\OtherTok{\textless{}{-}} \FunctionTok{stan\_model}\NormalTok{(}\AttributeTok{file=} \FunctionTok{here}\NormalTok{(}\StringTok{"Script"}\NormalTok{, }
                                      \StringTok{"Hevo"}\NormalTok{, }
                                      \StringTok{"MRHevo\_summarystats.stan"}\NormalTok{),}
                           \AttributeTok{model\_name=}\StringTok{"MRHevo.summarystats"}\NormalTok{, }
                           \AttributeTok{verbose=}\ConstantTok{FALSE}\NormalTok{,}
                           \AttributeTok{save\_dso =} \ConstantTok{TRUE}\NormalTok{,}
                           \AttributeTok{auto\_write =} \ConstantTok{TRUE}\NormalTok{)}

\NormalTok{get\_summary\_MR\_tib\_row }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(model\_list)\{}
  
  
  \CommentTok{\# Create output tibble in same format as Table 2/3 from}
  \CommentTok{\# Bowden et al}
\NormalTok{  output\_tib\_row }\OtherTok{\textless{}{-}} \FunctionTok{tibble}\NormalTok{(}\AttributeTok{N =} \FunctionTok{as.integer}\NormalTok{(),}
                           \AttributeTok{Prop\_Invalid =} \FunctionTok{as.double}\NormalTok{(),}
                           \AttributeTok{F\_stat =} \FunctionTok{as.double}\NormalTok{(),}
                           \AttributeTok{R2\_stat =} \FunctionTok{as.double}\NormalTok{(),}
                           \AttributeTok{WME\_Av =} \FunctionTok{as.double}\NormalTok{(),}
                           \AttributeTok{WME\_SE =} \FunctionTok{as.double}\NormalTok{(),}
                           \AttributeTok{WME\_Pos\_Rate =} \FunctionTok{as.double}\NormalTok{(),}
                           \AttributeTok{Hevo\_Av =} \FunctionTok{as.double}\NormalTok{(),}
                           \AttributeTok{Hevo\_SE =} \FunctionTok{as.double}\NormalTok{(),}
                           \AttributeTok{Hevo\_Pos\_Rate =} \FunctionTok{as.double}\NormalTok{())}
  
\NormalTok{  n\_datasets }\OtherTok{\textless{}{-}} \FunctionTok{length}\NormalTok{(model\_list)}
  
  \CommentTok{\# Create blank tibble to receive results of Weighted}
  \CommentTok{\# Median Estimator function from MR{-}Base}
  
\NormalTok{  results\_tib }\OtherTok{\textless{}{-}}  \FunctionTok{tibble}\NormalTok{(}\AttributeTok{WME\_est =} \FunctionTok{as.double}\NormalTok{(),}
                         \AttributeTok{WME\_se =} \FunctionTok{as.double}\NormalTok{(),}
                         \AttributeTok{WME\_pval =} \FunctionTok{as.double}\NormalTok{(),}
                         \AttributeTok{WME\_nsnp =} \FunctionTok{as.integer}\NormalTok{(),}
                         \AttributeTok{Hevo\_est =} \FunctionTok{as.double}\NormalTok{(),}
                         \AttributeTok{Hevo\_se =} \FunctionTok{as.double}\NormalTok{(),}
                         \AttributeTok{Hevo\_sd =} \FunctionTok{as.double}\NormalTok{(),}
                         \AttributeTok{Hevo\_est\_lower\_CI =} \FunctionTok{as.double}\NormalTok{(),}
                         \AttributeTok{Hevo\_est\_upper\_CI =} \FunctionTok{as.double}\NormalTok{(),}
                         \AttributeTok{Hevo\_causal\_detected =} \FunctionTok{as.logical}\NormalTok{()}
\NormalTok{  )}
  
  
  \CommentTok{\# Run WME and MR{-}Hevo for each dataset }
  \ControlFlowTok{for}\NormalTok{(dataset }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\NormalTok{n\_datasets)\{}
    
    \CommentTok{\# Stored as individual vectors for MR{-}Hevo/RStan {-} not}
    \CommentTok{\# Tidyverse compatible}
\NormalTok{    coeff\_G\_X\_vect }\OtherTok{\textless{}{-}}\NormalTok{ model\_list[[dataset]]}\SpecialCharTok{$}\NormalTok{coeff\_G\_X}
\NormalTok{    coeff\_G\_Y\_vect }\OtherTok{\textless{}{-}}\NormalTok{ model\_list[[dataset]]}\SpecialCharTok{$}\NormalTok{coeff\_G\_Y}
\NormalTok{    coeff\_G\_X\_SE\_vect }\OtherTok{\textless{}{-}}\NormalTok{ model\_list[[dataset]]}\SpecialCharTok{$}\NormalTok{coeff\_G\_X\_SE}
\NormalTok{    coeff\_G\_Y\_SE\_vect }\OtherTok{\textless{}{-}}\NormalTok{ model\_list[[dataset]]}\SpecialCharTok{$}\NormalTok{coeff\_G\_Y\_SE}
\NormalTok{    prop\_invalid }\OtherTok{\textless{}{-}} \FunctionTok{min}\NormalTok{(model\_list[[dataset]]}\SpecialCharTok{$}\NormalTok{prop\_invalid)}
\NormalTok{    F\_stat }\OtherTok{\textless{}{-}} \FunctionTok{min}\NormalTok{(model\_list[[dataset]]}\SpecialCharTok{$}\NormalTok{F\_stat)}
\NormalTok{    R2\_stat }\OtherTok{\textless{}{-}} \FunctionTok{min}\NormalTok{(model\_list[[dataset]]}\SpecialCharTok{$}\NormalTok{R2\_stat)}
\NormalTok{    n\_instruments }\OtherTok{\textless{}{-}} \FunctionTok{max}\NormalTok{(model\_list[[dataset]]}\SpecialCharTok{$}\NormalTok{Instrument)}
\NormalTok{    n\_participants }\OtherTok{\textless{}{-}} \FunctionTok{min}\NormalTok{(model\_list[[dataset]]}\SpecialCharTok{$}\NormalTok{n\_participants)}
    
    
    \CommentTok{\# N.B. MR{-}Hevo terminology vs WME paper/other code:}
    \CommentTok{\# alpha = effects of instruments on exposure, i.e. coeff\_G\_X}
    \CommentTok{\# beta = pleiotropic effects of instruments on outcome, i.e. alpha in WME}
    \CommentTok{\# gamma = effects of instruments on outcome, i.e. coeff\_G\_Y}
    \CommentTok{\# theta = causal effect X on Y, i.e. b}
    
    \CommentTok{\# Results from weighted median estimator method}
\NormalTok{    WME\_results }\OtherTok{\textless{}{-}} \FunctionTok{mr\_weighted\_median}\NormalTok{(}\AttributeTok{b\_exp =}\NormalTok{ coeff\_G\_X\_vect,}
                                      \AttributeTok{b\_out =}\NormalTok{ coeff\_G\_Y\_vect,}
                                      \AttributeTok{se\_exp =}\NormalTok{ coeff\_G\_X\_SE\_vect,}
                                      \AttributeTok{se\_out =}\NormalTok{ coeff\_G\_Y\_SE\_vect,}
                                      \AttributeTok{parameters =} \FunctionTok{list}\NormalTok{(}\AttributeTok{nboot =} \DecValTok{1000}\NormalTok{))}
    
    \CommentTok{\# Results from MR{-}Hevo method}
\NormalTok{    Hevo\_results}\OtherTok{\textless{}{-}} \FunctionTok{run\_mrhevo.sstats}\NormalTok{(}\AttributeTok{alpha\_hat =}\NormalTok{ coeff\_G\_X\_vect,}
                                    \AttributeTok{se.alpha\_hat =}\NormalTok{ coeff\_G\_X\_SE\_vect,}
                                    \AttributeTok{gamma\_hat =}\NormalTok{ coeff\_G\_Y\_vect,}
                                    \AttributeTok{se.gamma\_hat =}\NormalTok{ coeff\_G\_Y\_SE\_vect) }\SpecialCharTok{\%\textgreater{}\%}
     \FunctionTok{summary}\NormalTok{()}
    
    
    \CommentTok{\# Extract WME Results}
\NormalTok{    results\_tib[dataset, ]}\SpecialCharTok{$}\NormalTok{WME\_est }\OtherTok{\textless{}{-}}\NormalTok{ WME\_results}\SpecialCharTok{$}\NormalTok{b}
\NormalTok{    results\_tib[dataset, ]}\SpecialCharTok{$}\NormalTok{WME\_se }\OtherTok{\textless{}{-}}\NormalTok{ WME\_results}\SpecialCharTok{$}\NormalTok{se}
\NormalTok{    results\_tib[dataset, ]}\SpecialCharTok{$}\NormalTok{WME\_pval }\OtherTok{\textless{}{-}}\NormalTok{ WME\_results}\SpecialCharTok{$}\NormalTok{pval}
\NormalTok{    results\_tib[dataset, ]}\SpecialCharTok{$}\NormalTok{WME\_nsnp }\OtherTok{\textless{}{-}}\NormalTok{ WME\_results}\SpecialCharTok{$}\NormalTok{nsnp}
    
    \CommentTok{\# Extract MR{-}Hevo Results}
\NormalTok{    results\_tib[dataset, ]}\SpecialCharTok{$}\NormalTok{Hevo\_est }\OtherTok{\textless{}{-}}\NormalTok{ Hevo\_results}\SpecialCharTok{$}\NormalTok{summary[}\StringTok{"theta"}\NormalTok{,}\StringTok{"mean"}\NormalTok{]}
\NormalTok{    results\_tib[dataset, ]}\SpecialCharTok{$}\NormalTok{Hevo\_se }\OtherTok{\textless{}{-}}\NormalTok{ Hevo\_results}\SpecialCharTok{$}\NormalTok{summary[}\StringTok{"theta"}\NormalTok{,}\StringTok{"se\_mean"}\NormalTok{]}
\NormalTok{    results\_tib[dataset, ]}\SpecialCharTok{$}\NormalTok{Hevo\_sd }\OtherTok{\textless{}{-}}\NormalTok{ Hevo\_results}\SpecialCharTok{$}\NormalTok{summary[}\StringTok{"theta"}\NormalTok{,}\StringTok{"sd"}\NormalTok{]}
\NormalTok{    results\_tib[dataset, ]}\SpecialCharTok{$}\NormalTok{Hevo\_est\_lower\_CI }\OtherTok{\textless{}{-}}\NormalTok{ Hevo\_results}\SpecialCharTok{$}\NormalTok{summary[}\StringTok{"theta"}\NormalTok{,}\StringTok{"2.5\%"}\NormalTok{]}
\NormalTok{    results\_tib[dataset, ]}\SpecialCharTok{$}\NormalTok{Hevo\_est\_upper\_CI }\OtherTok{\textless{}{-}}\NormalTok{ Hevo\_results}\SpecialCharTok{$}\NormalTok{summary[}\StringTok{"theta"}\NormalTok{,}\StringTok{"97.5\%"}\NormalTok{]}
    
\NormalTok{  \}}
  
  \CommentTok{\# Add causality Boolean to MR{-}Hevo}
\NormalTok{  results\_tib }\OtherTok{\textless{}{-}}\NormalTok{ results\_tib }\SpecialCharTok{\%\textgreater{}\%}
   \FunctionTok{mutate}\NormalTok{(}\AttributeTok{Hevo\_est\_causal\_detected =}\NormalTok{ (Hevo\_est\_lower\_CI }\SpecialCharTok{\textgreater{}} \DecValTok{0}  \SpecialCharTok{|}\NormalTok{ Hevo\_est\_upper\_CI }\SpecialCharTok{\textless{}} \DecValTok{0}\NormalTok{))}
  
  
\NormalTok{  output\_tib\_row }\OtherTok{\textless{}{-}}\NormalTok{ results\_tib }\SpecialCharTok{\%\textgreater{}\%} 
    \FunctionTok{summarise}\NormalTok{(}\AttributeTok{N =}\NormalTok{ n\_participants,}
              \AttributeTok{Prop\_Invalid =}\NormalTok{ prop\_invalid,}
              \AttributeTok{F\_stat =} \FunctionTok{mean}\NormalTok{(F\_stat),}
              \AttributeTok{R2\_stat =} \FunctionTok{mean}\NormalTok{(R2\_stat),}
              \AttributeTok{WME\_Av =} \FunctionTok{mean}\NormalTok{(WME\_est),}
              \AttributeTok{WME\_SE =} \FunctionTok{mean}\NormalTok{(WME\_se),}
              \AttributeTok{WME\_Pos\_Rate =} \FunctionTok{length}\NormalTok{(WME\_pval[WME\_pval }\SpecialCharTok{\textless{}} \FloatTok{0.05}\NormalTok{]) }\SpecialCharTok{/}\NormalTok{ n\_datasets,}
              \AttributeTok{Hevo\_Av =} \FunctionTok{mean}\NormalTok{(Hevo\_est),}
              \AttributeTok{Hevo\_SE =} \FunctionTok{mean}\NormalTok{(Hevo\_se),}
              \AttributeTok{Hevo\_Lower\_CI =} \FunctionTok{mean}\NormalTok{(Hevo\_est\_lower\_CI),}
              \AttributeTok{Hevo\_Upper\_CI =} \FunctionTok{mean}\NormalTok{(Hevo\_est\_upper\_CI),}
              \AttributeTok{Hevo\_Pos\_Rate =} \FunctionTok{sum}\NormalTok{(Hevo\_est\_causal\_detected) }\SpecialCharTok{/}\NormalTok{ n\_datasets}
\NormalTok{    ) }\SpecialCharTok{\%\textgreater{}\%} 
    \FunctionTok{mutate}\NormalTok{(}\FunctionTok{across}\NormalTok{(}\FunctionTok{where}\NormalTok{(is.double), round, }\DecValTok{3}\NormalTok{))}
  
  \FunctionTok{return}\NormalTok{(output\_tib\_row)}
  
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{test\_tib\_summ\_MR\_data }\OtherTok{\textless{}{-}}  \FunctionTok{get\_simulated\_MR\_data}\NormalTok{(}\AttributeTok{n\_participants =} \DecValTok{10000}\NormalTok{,}
                                                \AttributeTok{n\_instruments =} \DecValTok{25}\NormalTok{,}
                                                \AttributeTok{n\_datasets =} \DecValTok{2}\NormalTok{,}
                                                \AttributeTok{prop\_invalid =} \FloatTok{0.1}\NormalTok{,}
                                                \AttributeTok{beta\_val =} \FloatTok{0.1}\NormalTok{,}
                                                \AttributeTok{causal\_effect =} \ConstantTok{TRUE}\NormalTok{,}
                                                \AttributeTok{rand\_error =} \ConstantTok{TRUE}\NormalTok{,}
                                                \AttributeTok{two\_sample =} \ConstantTok{TRUE}\NormalTok{,}
                                                \AttributeTok{balanced\_pleio =} \ConstantTok{TRUE}\NormalTok{,}
                                                \AttributeTok{InSIDE\_satisfied =} \ConstantTok{TRUE}\NormalTok{)}

\NormalTok{test\_tib\_summ\_MR\_models }\OtherTok{\textless{}{-}} \FunctionTok{get\_models}\NormalTok{(test\_tib\_summ\_MR\_data)}

\NormalTok{test\_tib\_summ\_MR\_row }\OtherTok{\textless{}{-}} \FunctionTok{get\_summary\_MR\_tib\_row}\NormalTok{(test\_tib\_summ\_MR\_models)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## CHECKING DATA AND PREPROCESSING FOR MODEL 'MRHevo.summarystats' NOW.
## 
## COMPILING MODEL 'MRHevo.summarystats' NOW.
## 
## STARTING SAMPLER FOR MODEL 'MRHevo.summarystats' NOW.
\end{verbatim}

\begin{verbatim}
## 
## CHECKING DATA AND PREPROCESSING FOR MODEL 'MRHevo.summarystats' NOW.
## 
## COMPILING MODEL 'MRHevo.summarystats' NOW.
## 
## STARTING SAMPLER FOR MODEL 'MRHevo.summarystats' NOW.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{test\_tib\_summ\_MR\_row }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{kable}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{kable\_styling}\NormalTok{(}\AttributeTok{latex\_options=}\StringTok{"scale\_down"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{table}
\centering
\resizebox{\ifdim\width>\linewidth\linewidth\else\width\fi}{!}{
\begin{tabular}{r|r|r|r|r|r|r|r|r|r|r|r}
\hline
N & Prop\_Invalid & F\_stat & R2\_stat & WME\_Av & WME\_SE & WME\_Pos\_Rate & Hevo\_Av & Hevo\_SE & Hevo\_Lower\_CI & Hevo\_Upper\_CI & Hevo\_Pos\_Rate\\
\hline
10000 & 0.1 & 14.739 & 0.036 & 0.091 & 0.09 & 0.5 & 0.121 & 0.001 & -0.052 & 0.302 & 0\\
\hline
\end{tabular}}
\end{table}

\newpage

\section{Appendix: R Packages Used}\label{appendix-pkg}

\subsection{Package Citations}\label{package-citations}

This work was completed using R version 4.4.3\textsuperscript{\citeproc{ref-base}{19}} with the following R packages: acronymsdown v. 0.11.1\textsuperscript{\citeproc{ref-acronymsdown}{26}}, bookdown v. 0.43\textsuperscript{\citeproc{ref-bookdown2016}{27},\citeproc{ref-bookdown2025}{28}}, car v. 3.1.3\textsuperscript{\citeproc{ref-car}{29}}, cowplot v. 1.1.3\textsuperscript{\citeproc{ref-cowplot}{30}}, crayon v. 1.5.3\textsuperscript{\citeproc{ref-crayon}{31}}, devtools v. 2.4.5\textsuperscript{\citeproc{ref-devtools}{32}}, ggdag v. 0.2.13\textsuperscript{\citeproc{ref-ggdag}{33}}, gghighlight v. 0.4.1\textsuperscript{\citeproc{ref-gghighlight}{34}}, grateful v. 0.2.12\textsuperscript{\citeproc{ref-grateful}{35}}, grid v. 4.4.3\textsuperscript{\citeproc{ref-grid}{36}}, here v. 1.0.1\textsuperscript{\citeproc{ref-here}{37}}, infer v. 1.0.8\textsuperscript{\citeproc{ref-infer}{38}}, kableExtra v. 1.4.0\textsuperscript{\citeproc{ref-kableExtra}{39}}, knitr v. 1.50\textsuperscript{\citeproc{ref-knitr2014}{40}--\citeproc{ref-knitr2025}{42}}, matrixStats v. 1.5.0\textsuperscript{\citeproc{ref-matrixStats}{43}}, medicaldata v. 0.2.0\textsuperscript{\citeproc{ref-medicaldata}{25}}, parallel v. 4.4.3\textsuperscript{\citeproc{ref-parallel}{44}}, rmarkdown v. 2.29\textsuperscript{\citeproc{ref-rmarkdown2018}{45}--\citeproc{ref-rmarkdown2024}{47}}, rstan v. 2.32.7\textsuperscript{\citeproc{ref-rstan}{48}}, tidyverse v. 2.0.0\textsuperscript{\citeproc{ref-tidyverse}{21}}, TwoSampleMR v. 0.6.16\textsuperscript{\citeproc{ref-TwoSampleMR2017}{49},\citeproc{ref-TwoSampleMR2018}{50}}, wordcountaddin v. 0.3.0.9000\textsuperscript{\citeproc{ref-wordcountaddin}{51}}.

\subsection{Session Information}\label{session-information}

\begin{verbatim}
##  setting  value
##  version  R version 4.4.3 (2025-02-28 ucrt)
##  os       Windows 11 x64 (build 26100)
##  system   x86_64, mingw32
##  ui       RTerm
##  language (EN)
##  collate  English_United Kingdom.utf8
##  ctype    English_United Kingdom.utf8
##  tz       Europe/London
##  date     2025-06-08
##  pandoc   3.4 @ C:/Program Files/RStudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)
##  quarto   NA @ C:\\PROGRA~1\\RStudio\\resources\\app\\bin\\quarto\\bin\\quarto.exe
\end{verbatim}

\begin{verbatim}
## # A tibble: 22 x 5
##    package      ondiskversion loadedversion date       source                   
##    <chr>        <chr>         <chr>         <chr>      <chr>                    
##  1 acronymsdown 0.11.1        0.11.1        2025-06-07 Github (rchaput/acronyms~
##  2 bookdown     0.43          0.43          2025-04-15 CRAN (R 4.4.3)           
##  3 cowplot      1.1.3         1.1.3         2024-01-22 CRAN (R 4.4.3)           
##  4 dplyr        1.1.4         1.1.4         2023-11-17 CRAN (R 4.4.3)           
##  5 forcats      1.0.0         1.0.0         2023-01-29 CRAN (R 4.4.3)           
##  6 gghighlight  0.4.1         0.4.1         2023-12-16 CRAN (R 4.4.3)           
##  7 ggplot2      3.5.2         3.5.2         2025-04-09 CRAN (R 4.4.3)           
##  8 grateful     0.2.12        0.2.12        2025-04-30 CRAN (R 4.4.3)           
##  9 here         1.0.1         1.0.1         2020-12-13 CRAN (R 4.4.3)           
## 10 infer        1.0.8         1.0.8         2025-04-14 CRAN (R 4.4.3)           
## 11 kableExtra   1.4.0         1.4.0         2024-01-24 CRAN (R 4.4.3)           
## 12 lubridate    1.9.4         1.9.4         2024-12-08 CRAN (R 4.4.3)           
## 13 medicaldata  0.2.0         0.2.0         2021-08-16 CRAN (R 4.4.3)           
## 14 purrr        1.0.4         1.0.4         2025-02-05 CRAN (R 4.4.3)           
## 15 readr        2.1.5         2.1.5         2024-01-10 CRAN (R 4.4.3)           
## 16 rstan        2.32.7        2.32.7        2025-03-10 CRAN (R 4.4.3)           
## 17 StanHeaders  2.32.10       2.32.10       2024-07-15 CRAN (R 4.4.3)           
## 18 stringr      1.5.1         1.5.1         2023-11-14 CRAN (R 4.4.3)           
## 19 tibble       3.2.1         3.2.1         2023-03-20 CRAN (R 4.4.3)           
## 20 tidyr        1.3.1         1.3.1         2024-01-24 CRAN (R 4.4.3)           
## 21 tidyverse    2.0.0         2.0.0         2023-02-22 CRAN (R 4.4.3)           
## 22 TwoSampleMR  0.6.16        0.6.16        2025-06-05 https://mrcieu.r-univers~
\end{verbatim}

\newpage

\end{document}
