---
title: "Simulations"
output: 
  pdf_document:
    latex_engine: xelatex
    dev: cairo_pdf
---


The data generating model used was from Appendix 3 of Bowden et al (ref), and was as follows


\begin{equation} 
  U_i = \sum^J_{j=1} \phi_jG_{ij} + \epsilon_i^U
\end{equation}


\begin{equation} 
  X_i = \sum^J_{j=1} \gamma_jG_{ij} + U_i + \epsilon_i^X
\end{equation}

\begin{equation} 
  Y_i = \sum^J_{j=1} \alpha_jG_{ij} + \beta X_i + U_i + \epsilon_i^Y
\end{equation}

for participants indexed by $i = 1, . . . , N$, and genetic instruments indexed by $j = 1, . . . , J$.
The error terms $\epsilon_i^U , \epsilon_i^X$ and $\epsilon_i^Y$ were each drawn independently from standard normal distributions. The genetic effects on the exposure γj are drawn from a uniform distribution between 0.03 and 0.1. Pleiotropic effects $\alpha_j$ and $\phi_j$ were set to zero if the
genetic instrument was a valid instrumental variable. Otherwise (with probability 0.1, 0.2, or 0.3):

1. In Scenario 1 (balanced pleiotropy, InSIDE satisfied), the $\alpha_j$ parameter was drawn from a uniform distribution between −0.2 and 0.2.

2. In Scenario 2 (directional pleiotropy, InSIDE satisfied), the $\alpha_j$ parameter was drawn from a uniform distribution between 0 and 0.2.

3. In Scenario 3 (directional pleiotropy, InSIDE not satisfied), the $\phi_j$ parameter was drawn from a uniform distribution between −0.2 and 0.2.


The causal effect of the exposure on the outcome was either $\beta X = 0$ (null causal effect) or $\beta X = 0.1$ (positive causal effect). A total of 10 000 simulated datasets were generated for sample sizes of N = 10 000 and 20 [sic] participants. Only the summary data, that is genetic associations with the exposure and with the outcome and their standard errors as estimated by univariate regression on the genetic instruments in turn,were used by the analysis methods. In the two-sample setting, data were generated on 2N participants, and genetic associations with the exposure were estimated in the first N participants, and genetic associations with the outcome in the second N participants. 


```{r setup, include=FALSE}

library(tidyverse)
library(TwoSampleMR)
library(mrhevo)
library(rstan)
library(here)
library(data.table)

```


```{r wme-model}

# Define function to create data generating model
# Arguments based on Bowden et al
wme_model_sim <- function(n_participants = as.integer(), 
                          n_instruments = as.integer(),
                          n_datasets = as.integer(),
                          prop_invalid = 0.1,
                          causal_effect = TRUE,
                          balanced_pleio = TRUE,
                          InSIDE_satisfied = TRUE,
                          rand_error = TRUE,              # remove random errors for testing
                          beta_val = 0.1,                 # size of causal effect
                          allele_freq_min = 0.01,         # frequency of effect allele
                          allele_freq_max = 0.99,
                          gamma_min = 0.03,               # size of pleiotropic effects on exposure
                          gamma_max = 0.1,
                          alpha_min = -0.2,               # size of pleiotropic effects on outcome
                          alpha_max = 0.2,
                          phi_min = -0.2,                 # size of additional pleiotropic effects
                          phi_max = 0.2){                 # when InSIDE not satisfied

  
  
  # Initialise blank lists to receive datasets for
  # each of:
  #     U (vector: unmeasured confounding exposures per participant), 
  #     X (vector: exposure:outcome associations estimated per participant) 
  #     Y (vector: gene:outcome association estimated per participant), 
  #     G (Matrices: Genotype data)
  #
  #     gamma (vector: pleiotropic effects of each instrument on exposure)
  #     alpha (vector: pleiotropic effects of each instrument on outcome)
  #     phi (vector: additional pleiotropic effects of each instrument when InSIDE assumption not satisfied)
  U_list <- list()
  X_list <- list()
  Y_list <- list()
  G_X_list <- list()
  G_Y_list <- list()
  
  gamma_list <- list()
  alpha_list <- list()
  phi_list <- list()
  beta_list <- list()
  prop_invalid_list <- list()
  
  
  # Assign features common to all datasets
  beta <- if_else(causal_effect == TRUE, # size of causal effect
                  beta_val,
                  0)
  
  
  # Create N datasets by simulating genotype matrices with
  # 1 row per participant, 1 column per genetic instrument
  # Use these to estimate U, X + Y
  
  for(n in 1:n_datasets){
    
    # Create error terms for U, X + Y per participant,
    # each drawn from standard normal distribution
    # unless random error turned off (for testing)
    
    ifelse(rand_error == TRUE,
           U_epsilon_vect <- rnorm(n = 2 * n_participants),
           U_epsilon_vect <- rep(0, 2 * n_participants))

    ifelse(rand_error == TRUE,
           X_epsilon_vect <- rnorm(n = n_participants),
           X_epsilon_vect <- rep(0, n_participants))
    
    ifelse(rand_error == TRUE,
           Y_epsilon_vect <- rnorm(n = n_participants),
           Y_epsilon_vect <- rep(0, n_participants))

    
    ## -- Create matrix of genotypes -- 
    ## 0 = reference, i.e. zero effect alleles, 
    ## 1 = 1 effect allele, 2 = 2 effect alleles 
    
    
    # Probability of effect allele set per dataset  
    # for each instrument, default value set at  
    # random between 0.01-0.99 (i.e. both effect +
    # reference are common alleles)
    allele_freq_vect <- runif(n = n_instruments,
                              min = allele_freq_min,
                              max = allele_freq_max)
    
    
    
    # Assign genotypes by sampling from binomial distribution
    # twice (as two alleles) per participant with probability
    # equal to frequency of effect allele
    # Create twice as many genotypes as participants in sample
    # to simulate 2 sample MR, i.e. first half used to estimate
    # Gene:Exposure, second half used to estimate Gene:Outcome
    
    # Matrix where columns are instruments, rows are participants
    # Values 0, 1 or 2
    
    
    G_mat <- matrix(rbinom(n = 2 * n_participants * n_instruments,
                           size = 2,
                           prob = rep(allele_freq_vect, 2 * n_participants)),
                    nrow = 2 * n_participants,
                    ncol = n_instruments,
                    byrow = TRUE)

        
    # Set which instruments invalid
    invalid_instrument_vect <- rbinom(n = n_instruments,
                                   size = 1, 
                                   prob = prop_invalid)
  

    # Set genetic effects of each instrument on the exposure,
    # drawn from uniform distribution, min/max as per Bowden 
    # et al
    gamma_vect <- runif(n = n_instruments,
                        min = gamma_min,
                        max = gamma_max)
  
    
    # Set pleiotropic effects on outcome, Scenarios and 
    # min/max from Bowden et al
    alpha_vect <- double() # Pleiotropic effects of instruments on outcome
    phi_vect <- double() # Pleiotropic effects of confounders on outcome
    
    for(j in 1:n_instruments){
      ifelse(invalid_instrument_vect[j] == FALSE,
             alpha_vect[j] <- 0,
             ifelse(balanced_pleio == TRUE,
                    alpha_vect[j] <- runif(n = n_instruments,
                                           min = alpha_min,
                                           max = alpha_max),
                    alpha_vect[j] <- runif(n = n_instruments,
                                           min = 0,
                                           max = alpha_max)
             )
      )
      
      # Assign default phi = 0 unless unbalanced pleiotropy & 
      # InSIDE assumption not satisfied & genetic instrument invalid
      if(balanced_pleio == FALSE & InSIDE_satisfied == FALSE){
        ifelse(invalid_instrument_vect[j] == FALSE,
               phi_vect[j] <- 0,
               phi_vect[j] <- runif(n = 1,
                                    min = phi_min,
                                    max = phi_max)
        )
        
      }
      else{
        phi_vect[j] <- 0
      }
    }
    
    # Create vectors of estimates for U, X and Y per individual,
    # i.e. Ui, Xi and Yi. Uses matrix inner product operator " %*%" 
    # https://stackoverflow.com/questions/22060515/the-r-operator 
    # http://matrixmultiplication.xyz/

    
    Ui_vect <-  G_mat %*% phi_vect + U_epsilon_vect
    
    Xi_vect <-  G_mat[1:n_participants, ] %*% gamma_vect + 
                Ui_vect[1:n_participants, ] + 
                X_epsilon_vect

    Yi_vect <-  G_mat[(n_participants+1):(2*n_participants), ] %*% alpha_vect + 
                beta * Xi_vect + 
                Ui_vect[(n_participants+1):(2*n_participants), ] + 
                Y_epsilon_vect
    
    
    # Add vectors of estimates from this dataset to lists of 
    # estimates from all datasets
    U_list[[n]] <- Ui_vect
    
    X_list[[n]] <- Xi_vect
    
    Y_list[[n]] <- Yi_vect
    
    G_X_list[[n]] <- G_mat[1:n_participants, ]
    
    #--- Test for Causal Effect ---#
    #G_Y_list[[n]] <- G_mat[1:n_participants, ]
    
    G_Y_list[[n]] <- G_mat[(n_participants+1):(2*n_participants), ]
    
    alpha_list[[n]] <- alpha_vect
    
    gamma_list[[n]] <- gamma_vect
    
    phi_list[[n]] <- phi_vect
    
    beta_list[[n]] <- beta
    
    prop_invalid_list[[n]] <- prop_invalid
  
    
  }
  
  
  # Combine all outputs to return
  combined_list <- list(U = U_list,         # Estimates 
                        X = X_list, 
                        Y = Y_list,
                        G_X = G_X_list,     # Genotypes of 1st sample
                        G_Y = G_Y_list,     # Genotypes of 2nd sample
                        alpha = alpha_list, # Actual values for validating simulation
                        gamma = gamma_list,
                        phi = phi_list,
                        beta = beta_list,
                        prop_invalid = prop_invalid_list
                        )
  
  return(combined_list)
  
}
```

```{r extract-models-XY}


# Create plotting tibble with Mean/SD X + Y grouped by
# Dataset + instrument
extract_models_XY <- function(sim){
  
  output_list <- list()
  
  # Create linear models per dataset to get coefficients
  # for gene:exposure association (coeff_G_X) and gene:outcome
  # association (coeff_G_Y)
  for(dataset in 1:length(sim$X)){
 
    X <- sim$X[[dataset]]
    Y <- sim$Y[[dataset]]
    Instruments_X <- sim$G_X[[dataset]]
    Instruments_Y <- sim$G_Y[[dataset]]
    
    alpha <- sim$alpha[[dataset]]
    gamma <- sim$gamma[[dataset]]
    phi <- sim$phi[[dataset]]
    beta <- sim$beta[[dataset]]
    prop_invalid <- sim$prop_invalid[[dataset]]

    
    # Model for gene:exposure
    X_lm <- lm(X ~ 0 + Instruments_X)
    coeff_G_X_vect <- coef(summary(X_lm))[1:(ncol(Instruments_X)), 1]
    SE_coeff_G_X_vect <- coef(summary(X_lm))[1:(ncol(Instruments_X)), 2]

    # Model for gene:outcome
    Y_lm <- lm(Y ~ 0 + Instruments_Y)
    coeff_G_Y_vect <- coef(summary(Y_lm))[1:(ncol(Instruments_Y)), 1] 
    SE_coeff_G_Y_vect <- coef(summary(Y_lm))[1:(ncol(Instruments_Y)), 2]

    output_list[[dataset]] <- as_tibble(list(dataset = dataset,
                                             Instrument = c(1:ncol(Instruments_X)),
                                             coeff_G_X = coeff_G_X_vect,
                                             coeff_G_X_SE = SE_coeff_G_X_vect,
                                             gamma = gamma,
                                             coeff_G_Y = coeff_G_Y_vect,
                                             coeff_G_Y_SE = SE_coeff_G_Y_vect,
                                             alpha = alpha,
                                             phi = phi,
                                             beta = beta,
                                             prop_invalid = prop_invalid),
                                        .name_repair = "unique")
  }
  
  return(output_list)

}  


```

A series of test plots were used to verify that data were simulated as intended under the various conditions required:

```{r test-tib, cache=TRUE}

# Check data produced in expected format
set.seed(1701)
sim_test_data_tib <- wme_model_sim(n_participants = 1000,
                                 n_instruments = 25,
                                 n_datasets = 2,
                                 prop_invalid = 0.3,
                                 rand_error = FALSE,
                                 causal_effect = TRUE,
                                 balanced_pleio = TRUE,
                                 InSIDE_satisfied = TRUE)

str(sim_test_data_tib)

```

```{r test-plot-prop-invalid, cache=TRUE}

# Check altering proportion of invalid instruments alters 
# proportion of instruments displaying pleiotropic effects
# N.B. cluster around alpha = 0 represents valid instruments with
# no pleiotropic effects

# 10% of instruments invalid
set.seed(1701)
sim_test_data_inval_0.1 <- wme_model_sim(n_participants = 100000,
                                     n_instruments = 25,
                                     n_datasets = 1,
                                     prop_invalid = 0.1,
                                     causal_effect = FALSE,
                                     balanced_pleio = TRUE,
                                     InSIDE_satisfied = TRUE)

# 50% of instruments invalid
set.seed(1701)
sim_test_data_inval_0.5 <- wme_model_sim(n_participants = 100000,
                                     n_instruments = 25,
                                     n_datasets = 1,
                                     prop_invalid = 0.5,
                                     causal_effect = FALSE,
                                     balanced_pleio = TRUE,
                                     InSIDE_satisfied = TRUE)


test_plot_tib_inval_0.1 <- extract_models_XY(sim_test_data_inval_0.1)[[1]]
test_plot_tib_inval_0.5 <- extract_models_XY(sim_test_data_inval_0.5)[[1]]

test_plot_tib_inval_0.1 %>% 
  select(alpha, coeff_G_Y) %>% 
  plot(.,
       ylab = "Observed Gene:Outcome Coefficient",
       main = "10% Invalid Instruments")

test_plot_tib_inval_0.5 %>% 
  select(alpha, coeff_G_Y) %>% 
  plot(.,
       ylab = "Observed Gene:Outcome Coefficient",
       main = "50% Invalid Instruments")



```

```{r test-plot-gamma-1, cache=TRUE}

# Check observed gene:exposure coefficients for each instrument
# (coeff_G_X) approximate true values (gamma) when a causal effect 
# is present & a large number of participants are included 
set.seed(1701)
sim_test_data_gamma_1 <- wme_model_sim(n_participants = 100,
                                       n_instruments = 25,
                                       n_datasets = 1,
                                       prop_invalid = 0.1,
                                       causal_effect = FALSE,
                                       rand_error = FALSE,
                                       balanced_pleio = TRUE,
                                       InSIDE_satisfied = TRUE)


test_plot_tib_gamma_1 <- extract_models_XY(sim_test_data_gamma_1)[[1]] 

test_plot_tib_gamma_1 %>% 
  select(gamma, coeff_G_X) %>% 
  plot(.,
       ylab = "Observed Gene:Exposure Coefficient")

```

```{r test-plot-alpha, cache=TRUE}

# Check gene:outcome coefficients (coeff_G_Y) approximate 
# pleiotropic effects (alphas) when no causal effect present
# N.B. cluster around alpha = 0 represents valid instruments with
# no pleiotropic effects

set.seed(1701)
sim_test_data_alpha <- wme_model_sim(n_participants = 10000,
                                     n_instruments = 25,
                                     n_datasets = 2,
                                     prop_invalid = 0.5,
                                     beta_val = 1,
                                     causal_effect = FALSE,
                                     balanced_pleio = TRUE,
                                     InSIDE_satisfied = TRUE,
                                     rand_error = FALSE)


test_plot_tib_alpha <- extract_models_XY(sim_test_data_alpha)[[2]]

test_plot_tib_alpha %>% 
  select(alpha, coeff_G_Y) %>% 
  plot(.,
       ylab = "Observed Gene:Outcome Coefficient")



```

```{r plot-GY-GX-fn}

plot_GY_GX <- function(model_tib, plot_title = as.character(NA)){
  
  model_tib %>% 
  mutate(Gradient = round(coefficients(lm(coeff_G_Y ~ 0 + coeff_G_X))[1], 5)) %>%
  ggplot() +
  aes(x = coeff_G_X, y = coeff_G_Y) +
  geom_point() +
  geom_abline(aes(intercept = 0, 
                  slope = Gradient),) +
  geom_text(aes(x = 0.075, # labels with gradient (causal effect estimate)
                y = 0.05, 
                label = paste0("\U03B2 == ", Gradient)), #beta
            hjust = 0,
            parse = TRUE) +
  annotate("text", 
           x = 0.075,      # add hat to beta
           y = 0.053, 
           label = paste("\U02C6"),
           hjust = -0.4,
           vjust = 0.9,
           parse = TRUE) +
  labs(title = plot_title,
       x = "Gene:Exposure Coefficient",
       y = "Gene:Outcome Coefficient") +
  xlim(0, 0.1) +
  ylim(-0.05, 0.06)
  
}


```

```{r test-plot-no-causal, cache=TRUE}

# No causal effect present
set.seed(1701)
sim_test_data_causal_0 <- wme_model_sim(n_participants = 10000,
                                        n_instruments = 1000,
                                        n_datasets = 1,
                                        prop_invalid = 0,
                                        causal_effect = FALSE,
                                        balanced_pleio = TRUE,
                                        InSIDE_satisfied = TRUE,
                                        rand_error = FALSE,
                                        beta_val = 0.1,                 # size of causal effect
                                        gamma_min = 0.03,               # size of pleiotropic effects on exposure
                                        gamma_max = 0.1,
                                        alpha_min = -0.2,               # size of pleiotropic effects on outcome
                                        alpha_max = 0.2)

test_plot_tib_causal_0 <- extract_models_XY(sim_test_data_causal_0)[[1]]

plot_GY_GX(test_plot_tib_causal_0, plot_title = "No Causal Effect")

```

```{r test-plot-causal, cache=TRUE}

# Check altering proportion of invalid instruments alters 
# proportion of instruments displaying pleiotropic effects
# N.B. cluster around alpha = 0 represents valid instruments with
# no pleiotropic effects


# Causal effect present
set.seed(1701)
sim_test_data_causal_1 <- wme_model_sim(n_participants = 1000,
                                        n_instruments = 100,
                                        n_datasets = 1,
                                        prop_invalid = 0,
                                        causal_effect = TRUE,
                                        balanced_pleio = TRUE,
                                        InSIDE_satisfied = TRUE,
                                        rand_error = FALSE,
                                        beta_val = 0.1,                 # size of causal effect
                                        gamma_min = 0.03,               # size of pleiotropic effects on exposure
                                        gamma_max = 0.1,
                                        alpha_min = -0.2,               # size of pleiotropic effects on outcome
                                        alpha_max = 0.2)

test_plot_tib_causal_1 <- extract_models_XY(sim_test_data_causal_1)[[1]]

plot_GY_GX(test_plot_tib_causal_1, plot_title = "Causal Effect Present")


```



```{r test-plot-phi, cache=TRUE}

# Check violating InSIDE assumption results in distorted 
# estimation of pleiotropic effects
# N.B. cluster around alpha = 0 represents valid instruments with
# no pleiotropic effects
set.seed(1701)
sim_test_data_phi_1 <- wme_model_sim(n_participants = 100000,
                                     n_instruments = 100,
                                     n_datasets = 1,
                                     prop_invalid = 0.3,
                                     causal_effect = FALSE,
                                     balanced_pleio = FALSE,
                                     InSIDE_satisfied = FALSE)

set.seed(1701)
sim_test_data_phi_0 <- wme_model_sim(n_participants = 100000,
                                     n_instruments = 100,
                                     n_datasets = 1,
                                     prop_invalid = 0.3,
                                     causal_effect = FALSE,
                                     balanced_pleio = FALSE,
                                     InSIDE_satisfied = TRUE)


test_plot_tib_phi_1 <- extract_models_XY(sim_test_data_phi_1)[[1]]
test_plot_tib_phi_0 <- extract_models_XY(sim_test_data_phi_0)[[1]]

test_plot_tib_phi_1 %>% 
  select(alpha, coeff_G_Y) %>% 
  plot(., 
       main = "InSIDE Violated",
       ylab = "Gene:Outcome Coefficient")

test_plot_tib_phi_0 %>% 
  select(alpha, coeff_G_Y) %>% 
  plot(.,
       main = "InSIDE Not Violated",
       ylab = "Gene:Outcome Coefficient")

#phi on y, not alpha


```
```{r test-plot-phi-2, cache=TRUE}

# Check violating InSIDE assumption results in distorted 
# estimation of pleiotropic effects
# N.B. cluster around alpha = 0 represents valid instruments with
# no pleiotropic effects
set.seed(1701)
sim_test_data_phi_T <- wme_model_sim(n_participants = 100000,
                                     n_instruments = 100,
                                     n_datasets = 1,
                                     prop_invalid = 0.3,
                                     causal_effect = FALSE,
                                     balanced_pleio = FALSE,
                                     InSIDE_satisfied = FALSE)

set.seed(1701)
sim_test_data_phi_F <- wme_model_sim(n_participants = 100000,
                                     n_instruments = 100,
                                     n_datasets = 1,
                                     prop_invalid = 0.3,
                                     causal_effect = FALSE,
                                     balanced_pleio = FALSE,
                                     InSIDE_satisfied = TRUE)


test_plot_tib_phi_T <- extract_models_XY(sim_test_data_phi_T)[[1]]
test_plot_tib_phi_F <- extract_models_XY(sim_test_data_phi_F)[[1]]

test_plot_tib_phi_T %>% 
  select(phi, coeff_G_Y) %>% 
  plot(., 
       main = "InSIDE Violated",
       ylab = "Gene:Outcome Coefficient")

test_plot_tib_phi_F %>% 
  select(phi, coeff_G_Y) %>% 
  plot(.,
       main = "InSIDE Not Violated",
       ylab = "Gene:Outcome Coefficient")

#phi on y, not alpha


```


```{r summary-MR-tib-row-fn}

get_summary_MR_tib_row <- function(n_participants = 1000,          # update to 10,000 for final
                                   n_instruments = 25,
                                   n_datasets = 100,               # update to 10,000 for final
                                   prop_invalid = 0.1,
                                   causal_effect = TRUE,
                                   balanced_pleio = TRUE,
                                   InSIDE_satisfied = TRUE,
                                   rand_error = TRUE,
                                   beta_val = 0.1,
                                   allele_freq_min = 0.01,
                                   allele_freq_max = 0.99,
                                   gamma_min = 0.03,               
                                   gamma_max = 0.1,
                                   alpha_min = -0.2,               
                                   alpha_max = 0.2,
                                   phi_min = -0.2,                 
                                   phi_max = 0.2){        
  
  # Generate data as per Bowden models with helper function 
  dataset_list <- wme_model_sim(n_participants = n_participants,          
                                n_instruments = n_instruments,
                                n_datasets = n_datasets,               
                                prop_invalid = prop_invalid,
                                causal_effect = causal_effect,
                                balanced_pleio = balanced_pleio,
                                InSIDE_satisfied = InSIDE_satisfied,
                                rand_error = rand_error,              
                                beta_val = beta_val,                 
                                allele_freq_min = allele_freq_min,         
                                allele_freq_max = allele_freq_max,
                                gamma_min = gamma_min,                
                                gamma_max = gamma_max,
                                alpha_min = alpha_min,               
                                alpha_max = alpha_max,
                                phi_min = phi_min,                 
                                phi_max = phi_max)
  

  # Use data to create model per dataset
  model_list <- extract_models_XY(dataset_list)
  
  
  # Create output tibble in same format as Table 2/3 from
  # Bowden et al
  output_tib_row <- tibble(N = as.integer(),
                           Prop_Invalid = as.double(),
                           F_stat = as.double(),
                           R2_stat = as.double(),
                           WME_Av = as.double(),
                           WME_SE = as.double(),
                           WME_Pos_Rate = as.double(),
                           Hevo_Av = as.double(),
                           Hevo_SE = as.double(),
                           Hevo_Pos_Rate = as.double())
  
  # Get number of datasets for row
  N_datasets <-  length(model_list)
  
  output_tib_row$N <-  N_datasets
  
  # Create blank tibble to receive results of Weighted
  # Median Estimator function from MR-Base
  
  results_tib <-  tibble(WME_b = as.double(),
                         WME_se = as.double(),
                         WME_pval = as.double(),
                         WME_Q = as.double(),
                         WME_Q_df = as.double(),
                         WME_Q_pval = as.double(),
                         WME_nsnp = as.integer())
  
  
  # Run WME for each dataset 
  for(dataset in 1:N_datasets){
    
    WME_results <- mr_weighted_median(b_exp = model_list[[dataset]]$coeff_G_X,
                                      b_out = model_list[[dataset]]$coeff_G_Y,
                                      se_exp = model_list[[dataset]]$coeff_G_X_SE,
                                      se_out = model_list[[dataset]]$coeff_G_Y_SE)
    
    results_tib[dataset, ]$WME_b <- WME_results$b
    results_tib[dataset, ]$WME_se <- WME_results$se
    results_tib[dataset, ]$WME_pval <- WME_results$pval
    results_tib[dataset, ]$WME_Q <- WME_results$Q
    results_tib[dataset, ]$WME_Q_df <- WME_results$Q_df
    results_tib[dataset, ]$WME_Q_pval <- WME_results$Q_pval
    results_tib[dataset, ]$WME_nsnp <- WME_results$nsnp
    
    # N.B. MR-Hevo terminology vs WME paper/other code:
      # alpha = effects of instruments on exposure, i.e. coeff_G_X
      # beta = pleiotropic effects of instruments on outcome, i.e. alpha in WME
      # gamma = effects of instruments on outcome, i.e. coeff_G_Y
      # theta = causal effect X on Y, i.e. b
    
  }
  
  output_tib_row <- results_tib %>% 
    summarise(N = N_datasets,
              Prop_Invalid = prop_invalid,
              F_Stat = NA,
              R2_stat = NA,
              WME_Av = mean(WME_b),
              WME_SE = mean(WME_se),
              WME_Pos_Rate = length(WME_pval[WME_pval < 0.05])/N_datasets,
              Hevo_Av = NA,
              Hevo_SE = NA,
              Hevo_Pos_Rate = NA) %>% 
    mutate(across(where(is.double), round, 3))
  
  return(output_tib_row)
}

```

```{r hevo-test, cache=TRUE}

# Standard set-up for RStan
options(mc.cores = parallel::detectCores())
rstan_options(auto_write = TRUE)

#source("functions.mrhevo.R")

# use.sampling <- TRUE
# logistic <- TRUE
# priorsd_theta <- 1
# newrun <- FALSE

set.seed(1701)
sim_test_data_hevo <- wme_model_sim(n_participants = 10000,
                                     n_instruments = 25,
                                     n_datasets = 2,
                                     prop_invalid = 0.1,
                                     beta_val = 0.1,
                                     causal_effect = TRUE,
                                     balanced_pleio = TRUE,
                                     InSIDE_satisfied = TRUE,
                                     rand_error = TRUE)


test_plot_tib_hevo <- extract_models_XY(sim_test_data_hevo)[[2]]



str(sim_test_data_hevo)
str(test_plot_tib_hevo)

a_h <- test_plot_tib_hevo$coeff_G_X
se_a_h <- test_plot_tib_hevo$coeff_G_X_SE
g_h <- test_plot_tib_hevo$coeff_G_Y
se_g_h <- test_plot_tib_hevo$coeff_G_X_SE

#test_mod <- test_plot_tib_hevo %>% 
  #as.data.frame() %>% 
  run_mrhevo.sstats(alpha_hat = a_h,
                    se.alpha_hat = se_a_h,
                    gamma_hat = g_h,
                    se.gamma_hat = se_g_h,
                    fraction_pleio = 0.2,
                    model.dir = here(),
                    priorsd_theta = priorsd_theta)



```


```{r test-summary-MR-tib-row-fn, warning=FALSE, cache=TRUE}



set.seed(14101583)
get_summary_MR_tib_row(n_participants = 10000,
                       n_instruments = 25,
                       n_datasets = 100,
                       prop_invalid = 0.3,
                       beta_val = 0.1,
                       causal_effect = TRUE,
                       balanced_pleio = TRUE,
                       InSIDE_satisfied = TRUE)



```




